docs: TSS Rosters design proposal (#14045)
hashgraph/hedera-services: platform-sdk/docs/proposals/001-tss-rosters-design-proposal/proposal.md

> # TSS - Roster Proposal

Reviewer: My comments from the previous revision still apply.

PS. It would be nice not to replace file names so as to allow rendering of diffs between revisions. It may be time consuming and unreasonable to re-read the entire old and new file just to try and catch (and probably miss some) changes.

Author: sorry about the renaming, just thought to bring it up to compliance with the Roster design template.

Reviewer: Please see all the open conversations in this PR above. Here's the direct links:

https://github.com/hashgraph/hedera-services/pull/14045#discussion_r1653736679
https://github.com/hashgraph/hedera-services/pull/14045#discussion_r1653744591
https://github.com/hashgraph/hedera-services/pull/14045#discussion_r1653746653
https://github.com/hashgraph/hedera-services/pull/14045#discussion_r1653747683
https://github.com/hashgraph/hedera-services/pull/14045#discussion_r1653750604
https://github.com/hashgraph/hedera-services/pull/14045#discussion_r1653752584

Author: addressed these. If i miss any, please shout.

---

## Summary

This proposal outlines the design specification rosters to support Threshold Signature Scheme (TSS) block signing.
It details the roster's lifecycle, data storage, API, and necessary changes to associated components within the
platform.

| Metadata           | Entities                                      | 
|--------------------|-----------------------------------------------|
| Designers          | Kore, Anthony, Ed, Cody, Austin               |
| Functional Impacts | TSS, Genesis, Cryptography, Reconnect         |
| Related Proposals  | TSS-Library, TSS-Ledger-Id, TSS-Block-Signing |
| HIPS               | N/A                                           |

## Purpose and Context

The introduction of the Threshold Signature Scheme (TSS) necessitates a new mechanism for managing node participation in
consensus and block signing.
Roster, an immutable subset of the address book, will provide this mechanism, ensuring efficient and secure key
management for TSS operations.
This proposal attempts to provide a specification for the behavior of rosters starting from their creation from the

> Candidate Address Book (CAB) to their terminal state within the TSS specification.

Author: Could be worded better but by `their terminal state within the TSS specification`, I meant the end state of the Candidate Roster - either it's adopted thereby becoming an active roster, or it's never adopted, remaining a candidate roster until network shutdown, or thrown away.

> Candidate Address Book (CAB) to their terminal state within the TSS specification.

Author: added a sentence to clarify what that means.

network's nodes. It will be continuously updated by HAPI transactions, such as those that create, update, or delete
nodes.
The CAB is a snapshot of the Future Address Book (FAB), and it is created when the app decides to adopt a new address
book.

This proposal specifies that Roster(s) will be created from the CAB by the Hedera app and passed to the platform code.
The mechanism for doing so is detailed below.

### Requirements

- Immutability: Rosters must be immutable to protect the integrity of its computed hash.
- Efficient Storage: Roster data should be stored efficiently in the state to minimize overhead and unnecessary data
  copying.
- Clear API: A well-defined API should be provided for the app to create and submit candidate rosters, and for the
  platform to manage the lifecycle of submitted rosters.
- Components update: Platform components that currently rely on the use of the Address book (such as Reconnect, Network,
  Event Validation etc.) must be adapted to use rosters instead.
- Address book paradigm discontinued within the platform codebase, only use rosters instead

### Architecture and/or Components

> ###### Roster API

Author: Added a lifecycle diagram. Thanks!

The Hedera app is already responsible for managing the address book.
We propose that it continues to do so, and at adoption time, creates a candidate Roster object from the CAB and set it
in the state.

> The State will contain one map of rosters, keyed by roster’s hash - ```Map<Hash<Roster>, Roster>```.

Reviewer 2: Need to add the data format for rosters so that people can know what is in it and what is not.   Maybe we should put the protobuf definition here?

Reviewer 3: +1. Or a class diagram would be nice.

Author: Copied over the definition, but in a class format.

Author: added the roster data structure

Reviewer 2: After consideration, I think it should be the protobuf that will be saved in the state that we use as the data structure.

Author: added protobuf

> It is noteworthy that the roster must be immutable to guarantee the integrity of the computed hash.

Reviewer 3: A question that can be answered by sharing the roster data structure is: what prevents two rosters on different stages (Active Roster, previously Active, previous Candidate, current Candidate) from having the same hash? 
Is there a date in the data-strucutre?
I mean, if matching hashes are possible, and later, some operations automatically remove a current candidate if a newer one is submitted, something should prevent rosters with the same hash from being deleted if they are indeed 2 different rosters.

Author: That's a good point. I'd like to have it such that creating multiple rosters of the same hash should have no effect (we've already seen this exact roster and processed it before). I'll make this explicit (except I hear otherwise).

> This map of rosters will typically contain the current Active Roster, an optional previously Active Roster, the optional

Reviewer 4: What is the purpose of each of these rosters? I don't see why we need a previously active roster or the previous candidate roster

Reviewer 2: The previous active roster I assume is the same as the previous address book and used in PCES replay and validating events from the older software version.

Author: We haven't discussed possibly deleting previously active rosters yet. We've kept it around similar to how we currently keep previous address book. Do you think we should remove previously active rosters immediately upon adoption of a new one?

I think it's ok to delete previous candidate rosters immediately a new candidate roster is sent.

previous Candidate Roster (a roster created from a candidate address book), and an optional current Candidate Roster.
There will be new fields in PlatformState - `candidateRosterHash` and `activeRosterHash` - such that at adoption time,
the way to trigger the adoption of a roster will be by the client code inserting a roster in the roster map, alongside

> setting the `candidateRosterHash` field in the PlatformState.

Reviewer 2: Suppose the app has previously set the `candidateRosterHash` several rounds ago and they set it again several rounds later.  At the time it is set a second time, overriding the first time, how does the platform detect that it has been set again?

Author: If it's the exact same one sent before, the hash will exist in the roster map so we will not override it. I've clarified above.

The `activeRosterHash` will be private to the platform and will not be settable or modifiable from outside the platform.

> This indirection avoids moving data around in the merkle state which requires data to be copied into the block stream,

Reviewer 4: I understand what you're getting at here, but many others will not. I think you are trying to explain why a map will be used to store the rosters instead of a singleton.

Author: thank you. I'm not particularly clear on what Singleton means here but you're probably right. The way i understand the need for indirection has to do with having to do a bunch of moving around (to update the merkle state) if the candidate roster was part of the state itself.

Author: Nvm - found a notion page that explains it. I'll adapt to include the justification for map vs singleton (really struggling to adjust to the use of the word "singleton" to describe this type of state data but I conceed it's way too late to rename it)

which is computationally expensive.
Another benefit of this approach is that adoption trigger becomes very simple (app sets the roster) with delineated
responsibilities between the app and the platform.

Some edge cases worth considering include:

1. multiple concurrent roster submission - has a potential for introducing race conditions. We will prevent this by
   guaranteeing immutability on rosters.
2. Size control on the map of rosters - we certainly don’t want this map to grow infinitely so insertion of rosters will

>    be controlled, with clear rules for removal of unused rosters (e.g. the acceptance of a new Candidate Roster

Reviewer 2: How is this enforced?

Author: I think the way the map can keep growing is by rapidly adding candidate rosters so fast (likely) or adopting new rosters rapidly (unlikely). As the example says, the quickest way to prevent  this is to attack the likely scenario - remove the current candidate roster once a new one is hashed. It will also address @kelly's question about why we need previous candidate roster. I've updated this to make this point very clear.

   invalidates and removes existing one).

###### Roster API - Implementation

Implementation of some new TSS components has been proposed.

> See [link](https://www.notion.so/TSS-Platform-Architecture-04b15df371ba4b1d848360542e05a030?pvs=21).

Reviewer 3: yea, also switching between documents is not ideal.

> See [link](https://www.notion.so/TSS-Platform-Architecture-04b15df371ba4b1d848360542e05a030?pvs=21).

Author: Good point visibility. Guess we have to duplicate that doc!

It is **important** to note that before the TSS protocol is implemented and launched, the Platform will adopt a
Candidate Roster during any software upgrade if the Candidate Roster is present in the state. Only after the TSS
protocol is launched will we be able to add an extra condition to verify that the Candidate Roster is complete, meaning
that it has enough TSS key material and votes accumulated.

### Core Behaviors, in summary

- Roster Creation: App will create a Candidate Roster from the Candidate Address Book (CAB) and set it in the state.
- Roster Submission: App will trigger roster adoption by setting the `candidateRosterHash` field in the PlatformState.
- Roster Adoption: The platform will adopt the last submitted Candidate Roster when it is ready, based on the TSS key

>   material and vote accumulation.

Reviewer 2: There is no TSS key material and votes in this proposal.   I think you should make reference to the TSS-Ledger-Id proposal once in the summary or context sections and just tailer the rest of the proposal to the TSS-Roster proposal.   It is getting confusing to know what the TSS-Roster proposal is proposing when it constantly references changes that are outside of its scope.

Author: agreed. Removed

- Roster Replacement: If a new CAB is submitted before the previous one is adopted, the corresponding new Candidate
  Roster will replace the old one.

> 

Reviewer 2: It would be good to show the life cycle changes in a diagram between before and after this proposal change.

Author: introduced a lifecycle diagram.

### Public API

A new method will be added to the platform API to allow the App submit a Candidate Roster to the platform:

```java
//in State

> void submitCandidateRoster(@NonNull final Roster candidateRoster);

Reviewer 4: i think this should just be `setCandidateRoster(...)` because it will be a simple setter method. The state is a shared data structure. There should be not logic in this method.

Reviewer 4: Need to include protobuf for the state being proposed.

Author: Renamed. I've added the proposed Roster's data structure , is that sufficient?

Reviewer 4: Yes, the protobuf is sufficient since the methods to set and access the data in the protobuf are automatically generated.

```

### Configuration

The `config.txt` file will no longer be used for storing the address book in existing networks. It will only be used for
the genesis of new networks.

### Data storage

> All the data related to Active and Candidate Roster, including the key material being accumulated via TSS messages, will

Reviewer 2: no need to reference the TSS messages as that is outside the scope of this proposal.

Author: removed.

> always be stored in the State.

Reviewer 2: Should we include the protobuf of the singleton leaf that is storing the rosters?

There will not be any other separate artifacts stored elsewhere (e.g., directly on disk.).

> The only artifact that will continue to be stored on disk separately from the State is the `config.txt`, which, as

Reviewer 2: This is not the only artifact.  There are also the public and private RSA key files.   You should explain how this change also effects how the public and private keys are managed, if changed at all.   (Public key lifecycle is changed.)

explained in *Bootstrapping Genesis for a brand-new network* section below, will only ever be used once during a genesis
of a brand-new network. The `config.txt` file, or at least its part that describes the address book components, will
never be used in the life cycle of an existing network after that.

### Startup procedure (pseudo-code)

The pseudocode for the startup procedure will look as follows:

```java
if(a State
exists on
disk){

loadStateFromDisk();
}else{
roster =

loadFromConfigTxt();

// Intall the roster as both Active Roster and Candidate Roster in the new state:
createEmptyState(roster);

> // Note that this Active Roster is missing TSS key material.

Reviewer 2: Mentioning key material here is going to confuse people not familiar with the other effort.

Author: removed.

// So the node will only be able to sign using its RSA key, w/o any TSS.
isGenesis =true;
        }

        if(
the Statehas
Candidate Roster){
        // Check if this is a software upgrade, and if the TSS protocol
        // has been launched already, then also check
        // if Candidate Roster has enough TSS key material/signatures.
        // Note that we switch to Candidate Roster during a software upgrade unconditionally
        // until the TSS protocol is actually launched.
        if(isSoftwareUpgrade /* && Candidate Roster is complete */){

// This MUST be performed under `isSoftwareUpgrade` to ensure that
// the entire network is being restarted, and so every node adopts the Candidate Roster,
// and hence no ISSes happen.
// Modify the state and put Candidate Roster into Active Roster, effectively clearing the Candidate Roster.

> makeCRtheAR();

Reviewer 2: How does this change impact the `AddressBookInitializer`?

> // May make a record of the previous Active Roster if necessary (e.g. for PCES replay)

Reviewer 2: Under what conditions is this a `May`?  Why not always?

// Note that the previous Active Roster may lack TSS key material
// if this is genesis, or if this is a TSS upgrade of an existing network.
    }

	/*

> 	// This block will be uncommented once the TSS protocol is implemented

Reviewer 2: I think talking about the conditions relevant in the TSS-Ledger-Id proposal  is confusing and not part of the scope of this proposal.   I'm not going to comment on it any more, but it applies to the rest of the comments in the pseudo code.

Author: @anthony-swirldslabs, ok if we just remove all references to `Key Material` from this proposal? Seems to be a common source of confusion.

Author: I've removed all references to `Key Material`, to be included in `TSS Ledger ID` proposal if required.

Reviewer: Yes, that sounds good to me. The references exist because the Roster design issue explicitly requested answers to questions related to the Roster life-cycle in TSS world. But I'm good if we move those references elsewhere.

	if (Candidate Roster is not complete) {
		// Need to populate the Candidate Roster with TSS key material and signatures.
		// Call the "TSS State Observer" to make it initiate the TSS protocol.
		// See the Detecting a new Candidate Roster (aka the new “Platform API”) section below.
		callTSSStateObserverToStartTSSProtocol();
		// NOTE: processing of TSS messages will accumulate TSS key material
		// in Candidate Roster and will emit a percentage metric indicating the readiness
		// of the Candidate Roster for adoption. Once the metric is at 100%, this shows
		// that the network can be restarted in order to adopt the Candidate Roster.
	}
	*/
            }

            if(
the State
has no
Active Roster){

// This should never happen
throwFatalErrorAndShutdown();
}

// Check if the Active Roster is TSS-enabled
        if(
Active Roster
is not
TSS-enabled){
  /*
  // This block will be uncommented once the TSS protocol is implemented
	if (isGenesis) {
		rejectAnythingButTSSMessages = true;
		// Note that a TSS upgrade of an existing network

> 		// shouldn't disable non-TSS messages processing.	

Reviewer 3: I don't follow how `rejectAnythingButTSSMessages` is set to true but there is a note about not disabling non-tss messages processing

Reviewer: This flag is set under the `if (isGenesis)` condition. An existing network is past its genesis, and therefore the `isGenesis` will be false for existing networks. And therefore this `if()` branch won't execute, leaving the flag as `false`. The comment simply emphasizes this fact for clarity.

Reviewer 3: Ok I understand, thanks for the explanation

        // above. So it will become TSS-enabled eventually, too, indicating
        // the readiness for a TSS upgrade via the metric mentioned above.
        // Note that until a Candidate Roster is installed in the state, the existing network
        // will simply continue to operate as usual with its regular RSA keys.
        }

// At this point the Active Roster in the state is what we'll be using as a roster.
// If an incomplete Candidate Roster exists, we've started exchanging TSS messages above.
// If it's genesis, we've disabled processing anything but TSS messages above.
// If Active Roster is already TSS-enabled (a mature network running after the TSS
// upgrade), then the node will be able to sign using TSS.
// If Active Roster isn't TSS-enabled yet, then the node will sign using its RSA key only,
// just as it does today.
```

### Bootstrapping Genesis for a brand new network

This proposal already creates a new interface for the services to install a Candidate Roster—via a Platform API, or
similar—which ultimately stores the Candidate Roster in a designated location in the State.

> The Startup procedure described above consumes the Candidate Roster from that location and starts its processing or

Reviewer 2: There's a complication here that the above pseudo code is not factoring in.  

If we don't do inversion of control to where Services parses `config.txt` then we have to explain how services gets the fields they need from config.txt.   Rosters do not have `memo` field which currently contains the Hedera account number for the node.  

It's easier to just do inversion of control and have the App set a genesis roster.  That way Services can get whatever they need from config.txt

adoption based on its readiness.

The `config.txt` will only ever be used literally once—when a brand new network goes through its genesis.
The `config.txt` will never be used again in the network life cycle.

Given that:

1. we need a special mechanism for supplying a genesis roster to a brand new network, and
2. this mechanism is only ever used once during genesis, and
3. we already have a support for `config.txt` in the platform, and
4. we already propose a public interface for Services to install Candidate Rosters in existing networks

there should be absolutely no need to introduce any additional mechanisms or interfaces between Platform and Services to

> supply a genesis roster.

Reviewer 2: this is false, see my last comment.    If we don't do inversion of control then we have to manage data that services needs in config.txt that is not stored in the roster.

> If we must, we may add a field to the `config.txt` to indicate the number of shares for each node.

Reviewer 2: The idea of `number of shares` has no place in this proposal.

  learner node uses the address book to select the teacher node, as well as remove all invalid signatures from the
  state. Both of these will need to be updated to use rosters instead.
- Networking. Most parts of the network code have been abstracted away from the address book as part of the Dynamic
  Address Book effort, so there should be minimal work left there.
- Event Validation. The event validation logic uses the address book to determine whether a given event has a valid
  signature. This will be updated to use rosters instead.
- Miscellaneous components. Other components that reference things stored in the address book (like consensus needing
  the node weight or the crypto module verifying keys) will need to be updated to use things stored in rosters instead.

### Test Plan

Some of the obvious test cases to be covered in the plan include validating one or more of the following scenarios:

1. New valid Candidate Roster created with no subsequent one sent by app. Verify accept.
2. New valid Candidate Roster created with subsequent one sent by app. Verify accept.

> 3. Invalid roster(s) sent. Verify reject.

Reviewer 2: What is the definition of `valid` roster?   That needs to be explained clearly and precisely in an earlier section.

Author: hmmm I'm not sure about including that actually. I think the presense of the protobuf/data structure implies what constitutes validity e.g. it must have >0 RosterEntry/ies, with each having non-zero weight with valid certificates etc. Do you think that's valuable information?

Author: Is something like this useful?

###### Roster Validity
In simple terms, the following constitutes a valid roster:
1. The roster must have at least one RosterEntry.
2. All RosterEntry/ies must have a non-zero weight.
3. All RosterEntry/ies must have a valid X509Certificate.
4. All RosterEntry/ies must have a valid PairingPublicKey.
5. All RosterEntry/ies must have at least one ServiceEndpoint.
6. All ServiceEndpoint/s must have a valid IP address, port, or domain name.
7. The roster must have a unique NodeId for each RosterEntry.

> 4. Empty roster(s) sent. Verify reject.

Reviewer 2: what does `sent` mean here? In what context is this being rejected?

Author: clarified.

5. Node Failures During Roster Change: What happens if nodes fails or disconnects during a roster change? Verify valid
   node successfully reconnects.
6. Concurrent Roster Updates: What if we make multiple roster updates concurrently? Verify no effect on adoption.

> 7. Rosters with nodes with zero or maximum shares. What happens?

Reviewer 2: Shares do not matter here.

8. Roster recovery? Node receives roster, crashes. Wakes up … ?

> 9. What testing do we need for genesis new network?

Reviewer 2: The answer to this question is?

Author: I don't know actually. One approach though is to treat the platform as a library, in which case, we need not do anything specific to testing a genesis new network - a network is a network, we just test that.

Author: removed & resolved

> 10. What testing do we need for an already new network with no previous TSS upgrade

Reviewer 2: What is an `already new network` and how is it different than genesis?

Author: removing this from here - belongs in TSS ledge ID tests imo.

> 11. What testing do we need for an already new network with previous TSS upgrade

Reviewer 2: I'm confused by this statement `already new network with previous TSS upgrade`

Author: wording could improve. We had been discussing a hybrid approach where there are two phases: 
1. Existing network that will use config.txt to bootstrap genesis
2. Subsequent network that will use TSS going forward

After the conversation to get rid of config.txt, we no longer need this and should just test TSS-enabled network instead. I've reworded it.

### Metrics

> We propose that some metrics be added to track the progress of the TSS key material and votes accumulation in the

Reviewer 2: specific metrics?  (to this proposal, not related to other proposals).

Author: removed reference to key material

Candidate Roster.
Others may include create timestamp of a Candidate Roster, and how long voting took.

> ### Implementation and Delivery Plan

Reviewer 2: How does this impact Services?  
How does this impact DevOps? 
Is a HIP required?

Reviewer 2: DevOps Impact: 
* When adding new nodes to existing networks, they need to be given a state after the network has upgraded.  The state contains the address book /roister with the new node in it.    Devops will no longer use config.txt on new nodes to existing networks.

Author: who provides the state to devops, services? Should we include that here?

Reviewer 2: I think we should indicate at a high level how Services and DevOps will need to change their processes and why.  

DevOps gets the state from disk when a snapshot is created.

Reviewer 5: With Block Node devops will get state from a Block Node, and any require filtering can be done there.
Not sure if it's directly relevant, but perhaps something worth considering.

Reviewer 2: I would hope that TSS-Roster is implemented and shipped before block node.  if not, we should adjust our language in this proposal to simply talk about the change in work flow in providing a state and not where the state comes from.    The state must have the node in its active roster.

Author: As per the project plan, TSS Roster delivery has to come before the rest of TSS spec.

> - Define Roster Data Structure and API: Design and implement the Roster class and the associated API methods.

Reviewer 2: That is what this proposal is for, define the data structure and API.   Please do that and add it to this proposal.

- Modify State Storage: Update the state to store the map of rosters and the candidateRosterHash and activeRosterHash
  fields.
- Implement Roster Adoption Logic: Develop the logic for the platform to adopt candidate rosters based on TSS readiness
  and voting.
- Update Reconnect Process: Modify the reconnect process to use rosters instead of address books.

> - Testing and Deployment: Conduct thorough testing of the new roster implementation and deploy.

Reviewer 2: What tests are relevant?  How do the existing JRS tests get modified?   Do we create new integration tests with a different framework than JRS?

This indirection avoids moving data around in the merkle state which requires data to be copied into the block stream,
which is computationally expensive.
Another benefit of this approach is that adoption trigger becomes simple (app sets the roster) with delineated
responsibilities between the app and the platform.

This map will not grow infinitely. Insertion of rosters will be controlled. The acceptance of a new Candidate Roster
will invalidate and remove the current candidate roster.

###### Roster Validity

In simple terms, the following constitutes a valid roster:

1. The roster must have at least one RosterEntry.
2. All RosterEntry/ies must have a non-zero weight.
3. All RosterEntry/ies must have a valid X509Certificate.
4. All RosterEntry/ies must have a valid PairingPublicKey.
5. All RosterEntry/ies must have at least one ServiceEndpoint.
6. All ServiceEndpoint/s must have a valid IP address, port, or domain name.
7. The roster must have a unique NodeId for each RosterEntry.

### Core Behaviors, in summary

- Roster Creation: App will create a Candidate Roster from the Candidate Address Book (CAB) and set it in the state.

> - Roster Submission: App will trigger roster adoption by setting the `candidateRosterHash` field in the PlatformState.

Author: Thanks, I've modified the wording. I've understood the state API to be PlatformState::setCandidateRoster(Roster roster), not SwirldState. Do you think that's wrong? It's important to clarify it.

> - Roster Adoption: The platform will adopt the last submitted Candidate Roster when it is ready, based on specifications

Author: There can be no adoption without voting, and the mechanism of voting is beyond this proposal (except we decide to include it in this work). I always understood the scope of this proposal to be the creation, submission, storage,  retrieval, and removal of the Roster, with the adoption process being out of scope. I'm not sure I have an answer for  how to adopt the rosters before TSS Ledger ID work is done.

> - Roster Adoption: The platform will adopt the last submitted Candidate Roster when it is ready, based on specifications

Reviewer 2: I think we cleared up the confusion.  The roster is always adopted on upgrade.

3. Decommissioning `config.txt`: This file does not have all the information that the Hedera Address Book needs (proxy
   endpoints, and in the future Block Nodes). It has verbose information for the platform. Its format could be better,
   and it stores the account number in the memo field. Upon the creation and adoption of rosters in the
   state, `config.txt` is no longer useful.
   The resolution is to offload this duty off the platform and allow Services to use whatever file format that suits
   them.

### Data storage

Most of the data related to Active and Candidate Roster will be stored in the State.
There will continue to be components used in creating the Roster - such as the private EC (elliptic curve) key, the RSA
private key and the signing X509Certificate certificate - that will be stored on disk.
However, it is up to Services to manage the lifecycle of these files, and not the platform.

There will not be any other separate artifacts stored elsewhere (e.g., directly on disk.).

### Startup procedure (pseudo-code)

The pseudocode for the startup procedure will look as follows:

```code
if(a State exists on disk) {
    loadStateFromDisk();
} else {

>     roster = loadFromConfigTxt();

Reviewer: It now does.

}

if(the State has Candidate Roster) {
    // Check if this is a software upgrade, and if the TSS protocol
    // has been launched already, then also check
    // if Candidate Roster has enough signatures.
    // Note that we switch to Candidate Roster during a software upgrade unconditionally
    // until the TSS protocol is actually launched.
    if(isSoftwareUpgrade /* && Candidate Roster is complete */) {

        // This MUST be performed under `isSoftwareUpgrade` to ensure that
        // the entire network is being restarted, and so every node adopts the Candidate Roster,
        // and hence no ISSes happen.
        // Modify the state and put Candidate Roster into Active Roster, effectively clearing the Candidate Roster.
        makeCRtheAR();
        // May make a record of the previous Active Roster if necessary (e.g. for PCES replay)
    }

	/*
	// This block will be uncommented once the TSS protocol is implemented
	if (Candidate Roster is not complete) {
		// Call the "TSS State Observer" to make it initiate the TSS protocol.
		// See the Detecting a new Candidate Roster (aka the new “Platform API”) section below.
		callTSSStateObserverToStartTSSProtocol();

> 		// NOTE: processing of TSS messages will emit a percentage metric indicating the readiness

Author: I did? Sorry about that. I've invited @anthony-swirldslabs to please take a look at this section.

> 		// NOTE: processing of TSS messages will emit a percentage metric indicating the readiness

Reviewer: Removed.

		// of the Candidate Roster for adoption. Once the metric is at 100%, this shows
		// that the network can be restarted in order to adopt the Candidate Roster.
	}
	*/
}

if(the State has no Active Roster) {

    // This should never happen
    throwFatalErrorAndShutdown();
}

// Check if the Active Roster is TSS-enabled

> if(Active Roster is not TSS-enabled) {

Reviewer: Removed.

            // shouldn't disable non-TSS messages processing.	
        }
    */

    // In Genesis, we've just disabled processing anything but TSS. So a new
    // network will eventually become TSS-enabled and only then will start
    // processing non-TSS events.
    // An existing, non-genesis network will continue to operate as before,
    // and if a Candidate Roster was present in the state, we enabled processing TSS messages
    // above. So it will become TSS-enabled eventually, too, indicating
    // the readiness for a TSS upgrade via the metric mentioned above.
    // Note that until a Candidate Roster is installed in the state, the existing network
    // will simply continue to operate as usual with its regular RSA keys.
}

// At this point the Active Roster in the state is what we'll be using as a roster.
// If an incomplete Candidate Roster exists, we've started exchanging TSS messages above.
// If it's genesis, we've disabled processing anything but TSS messages above.
// If Active Roster is already TSS-enabled (a mature network running after the TSS
// upgrade), then the node will be able to sign using TSS.

> // If Active Roster isn't TSS-enabled yet, then the node will sign using its RSA key only,

Reviewer: Removed.

It details the roster's lifecycle, data storage, API, and necessary changes to associated components within the
platform.

| Metadata           | Entities                                      | 
|--------------------|-----------------------------------------------|
| Designers          | Kore, Anthony, Ed, Cody, Austin               |
| Functional Impacts | TSS, Genesis, Cryptography, Reconnect         |
| Related Proposals  | TSS-Library, TSS-Ledger-Id, TSS-Block-Signing |
| HIPS               | N/A                                           |

## Purpose and Context

The introduction of the Threshold Signature Scheme (TSS) requires a new mechanism for managing node participation in
consensus and block signing.
Roster, an immutable subset of the address book, will provide this mechanism, ensuring efficient and secure key
management for TSS operations.
This proposal attempts to provide a specification for the behavior of rosters starting from their creation from the
Candidate Address Book (CAB) to their terminal state within the TSS specification.
A roster reaches a terminal state when it is either adopted by the platform or replaced by a new roster.

> The Future Address Book (FAB) is a dynamic list maintained within the App that reflects the desired future state of the

Author: I see what you're saying but I think it's important to make the proposal self-contained. That way, a user without context can pick it up and get some basic understanding of where this effort lies within the complex system that houses it. The CAB and FAB origin's summary here helps provide this context. Else, a reader could walk away wondering what CAB and FAB are and where they come from.

nodes.
The CAB is a snapshot of the Future Address Book (FAB), and it is created when the app decides to adopt a new address
book.

This proposal specifies that Roster(s) will be created from the CAB by the Hedera app and passed to the platform code.
The mechanism for doing so is detailed below.

### Requirements

- Immutability: Rosters must be immutable to protect the integrity of its computed hash.
- Efficient Storage: Roster data should be stored efficiently in the state to minimize overhead and unnecessary data
  copying.
- Clear API: A well-defined API should be provided for the app to create and submit candidate rosters, and for the
  platform to manage the lifecycle of submitted rosters.
- Components update: Platform components that currently rely on the use of the Address book (such as Reconnect, Network,
  Event Validation etc.) must be adapted to use rosters instead.
- Address book paradigm discontinued within the platform codebase, only use rosters instead

### Architecture and/or Components

###### Roster API

The Hedera app is already responsible for managing the address book.

> We propose that it continues to do so, and at adoption time, create a candidate Roster object from the CAB and set it in

Author: I thought it was helpful to inform the reader of the Roster's origin, though not in much detail.

the state.
The State will contain one map of rosters, keyed by roster’s hash - ```Map<Hash<Roster>, Roster>```.
A Roster has a data structure as follows:

```
class Roster {
    List~RosterEntry~ entries
}

class RosterEntry {
    NodeId nodeId
    long weight
    X509Certificate signingCert
    PairingPublicKey tssEcKey
    List~ServiceEndpoint~ gossipEndpoints
}

class ServiceEndpoint {
    bytes ipAddressV4
    int32 port
    string domain_name
}

Roster "1" *-- "many" RosterEntry
RosterEntry "1" *-- "many" ServiceEndpoint

```
It is noteworthy that the roster must be immutable to guarantee the integrity of the computed hash.
This map of rosters will typically contain the current Active Roster, an optional previously Active Roster, the optional

> previous Candidate Roster (a roster created from a candidate address book), and an optional current Candidate Roster.

Reviewer 6: Yes Ed, that's the idea ( off by 1 errors aside, which I'm personally not bothering to try to get straight... that's a job for the implementer :D )

> previous Candidate Roster (a roster created from a candidate address book), and an optional current Candidate Roster.

Author: I don't think I've documented how the round association will work (earlier idea was to include it in the Roster itself but that proposal had been witdrawn) so this is a valuable idea.

You're proposing an additional data structure - a queue of Roster hashes - where the index of the queue represents round numbers. One thing i'd add is to consider making the round vs roster hash relationship explicit. It will be easy to introduce bugs if implicit—wrong offset, queue pop at the wrong time or by the wrong thread, etc.).

Do we see ever having to do a round lookup (e.g. what roster was used x rounds ago)? 

I'd love to include this once we work out these tiny details. Thanks for raising it!

> previous Candidate Roster (a roster created from a candidate address book), and an optional current Candidate Roster.

Reviewer 6: To be clear, the round-number-as-index structure isn't something I'm advocating for right now. Rather, it's a plan compatible with full DAB, which utilizes a queue. The proposal for the short-term solution to use a queue, so that fewer changes are required when the transition to full DAB *is* made. The section in my original comment called "Pre full DAB" describes the short term proposal.

> One thing i'd add is to consider making the round vs roster hash relationship explicit. It will be easy to introduce bugs if implicit—wrong offset, queue pop at the wrong time or by the wrong thread, etc.).

We need to be very stingy with what we put into the state, if we can implicitly store this info via index, we should. If we're introducing off-by-one bugs in this area of the code, we can all pack up and go home anyway.

> Do we see ever having to do a round lookup (e.g. what roster was used x rounds ago)?

Yes, this is how we will validate events in a fully DAB system.

> previous Candidate Roster (a roster created from a candidate address book), and an optional current Candidate Roster.

Author: I'll setup a call to discuss this idea in detail.

> previous Candidate Roster (a roster created from a candidate address book), and an optional current Candidate Roster.

Author: Thanks all for the discussion last week.

FYI I've introduced a new diagram capturing this idea, and presented the pre-DAB implementation in this proposal. The full-DAB implementation will be implemented as part of dynamic address book.

  Roster will replace the old one.

### Public API

A new method will be added to the platform API to allow the App submit a Candidate Roster to the platform:

```java
//in State

void setCandidateRoster(@NonNull final Roster candidateRoster);
```

### Configuration

The `config.txt` file will no longer be used for storing the address book in existing networks. It will only be used for
the genesis of new networks.

### Data storage

All the data related to Active and Candidate Roster will be stored in the State.
There will not be any other separate artifacts stored elsewhere (e.g., directly on disk.).
The only artifact that will continue to be stored on disk separately from the State is the `config.txt`, which, as
explained in *Bootstrapping Genesis for a brand-new network* section below, will only ever be used once during a genesis
of a brand-new network. The `config.txt` file, or at least its part that describes the address book components, will
never be used in the life cycle of an existing network after that.

### Startup procedure (pseudo-code)

The pseudocode for the startup procedure will look as follows:

> ```java

Reviewer 6: IMO, this block of pseudocode isn't adding anything to the proposal. In reality, the stuff being done in this code block will happen inside various components, and on various threads. 

I'm generally in favor of expressing things with code, but in this case, I think attempting to render the process as a linear sequence of events gives the wrong impression, and won't really be useful when it comes time to actually implement this.

Author: @anthony-swirldslabs, thoughts?

Reviewer: This pseudocode was written to answer a few specific questions explicitly asked in the original problem statement at https://github.com/hashgraph/hedera-services/issues/13775 (see the tasks in the description of the issue.)

Also, I believe that the code as written would be running during the startup, on a single thread, in a single place, specifically around here: https://github.com/hashgraph/hedera-services/blob/d1f97ac28ec829cfb0e5d1f2ac5e10675a95af7c/platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/builder/PlatformBuilder.java#L505-L560

It does indeed call a few methods in other components, but 99% of this pseudocode itself belongs to this single startup method implementation directly.

However, if the problem statement has changed and those questions no longer require any answers in the design proposal, then the code can be removed.

Reviewer 2: The definition of `Public API` is not just what impacts the broader hedera ecosystem, but also anything that impacts Services.   In that sense, the `PlatformBuilder` and whole bootstrap process of instantiating a platform is public API when we do `inversion of control`.   Services needs to know where to set a `genesisRoster`.    Whatever process that is, it is `Public API` and needs to be documented clear enough for people in Services to call it properly.   That may mean that pseudocode here isn't enough, that we need to show the modified interfaces.

Author: is the interface definition under the `Roster Public API` sufficient? What else do you think would be helpful to specify here without going into concrete implementation detail?

Reviewer 2: If I was a services developer with only this document in hand and I was responsible for changing the services code to give the platform a genesis roster at startup, would I be able to do it correctly or figure it out?  How much time would it take me?

  the node weight or the crypto module verifying keys) will need to be updated to use things stored in rosters instead.

### Test Plan

Some of the obvious test cases to be covered in the plan include validating one or more of the following scenarios:

1. New valid Candidate Roster created with no subsequent one sent by app. Verify accept.
2. New valid Candidate Roster created with subsequent one sent by app. Verify accept.
3. Invalid roster(s) sent. Verify reject.
4. Empty roster(s) sent. Verify reject.
5. Node Failures During Roster Change: What happens if nodes fails or disconnects during a roster change? Verify valid
   node successfully reconnects.
6. Concurrent Roster Updates: What if we make multiple roster updates concurrently? Verify no effect on adoption.
7. Rosters with nodes with zero or maximum shares. What happens?
8. Roster recovery? Node receives roster, crashes. Wakes up … ?
9. What testing do we need for genesis new network?
10. What testing do we need for an already new network with no previous TSS upgrade
11. What testing do we need for an already new network with previous TSS upgrade

### Metrics

> We propose that some metrics be added such as `createTimestamp` of a Candidate Roster, and the time the roster was

Reviewer 6: Would this be a "timestamp? If we're talking about metrics, we would normally just increment a counter or something, and the metrics framework handles mapping that onto time.

Reviewer 2: Agreed, I think we would just count the number of times that `setCandidateRoster(Roster roster)` has been called.   The value resets to 0 or 1 on a node restart based on whether or not there is a candidate roster in the state on startup.

> We propose that some metrics be added such as `createTimestamp` of a Candidate Roster, and the time the roster was

Reviewer 6: Saying "some metrics be added", but only mentioning a single metric isn't very helpful. I'd say either only mention the one metric, or describe multiple metrics. The way it's worded, it sounds like there are other metrics in mind, that just aren't detailed here.

Reviewer 6: I don't think it's super important to force ourselves to come up with a bunch of metrics at the proposal stage. There's no problem noting down ones that we've thought of, but ancillary metrics can just be left as an implementation detail.

Author: agreed. I''ll' remove the metrics section. If we find a compelling metric, I'll bring it back.

The Future Address Book (FAB) is a dynamic list maintained within the App that reflects the desired future state of the
network's nodes. It will be continuously updated by HAPI transactions, such as those that create, update, or delete
nodes.
The CAB is a snapshot of the Future Address Book (FAB), and it is created when the app decides to adopt a new address
book.

This proposal specifies that one or more Roster(s) will be created from the CAB by the Hedera app and passed to the
platform code.
The mechanism for doing so is detailed below.

### Requirements

- Immutability: Rosters must be immutable to protect the integrity of its computed hash.
- Efficient Storage: Roster data should be stored efficiently in the state to minimize overhead and unnecessary data
  copying.
- Clear API: A well-defined API should be provided for the app to create and submit candidate rosters, and for the
  platform to manage the lifecycle of submitted rosters.
- Components update: Platform components that currently rely on the use of the Address book (such as Reconnect, Network,
  Event Validation etc.) must be adapted to use rosters instead.

> - Address book paradigm discontinued within the platform codebase, only use rosters instead

Author: No. I'm not sure I'm a fan of how we're branding "inversion of control" as it relates to the platform shifting responsibilities back to services. The terminology already defines a specific design principle. This is one implementation of that design principle. We should just keep it simple and say "services will provide X".

The logic applies to 'Singleton' terminology for the state. It means something else, we should stay clear of things with existing, popular meanings.
(I'm moaning, I know :) )

A Roster has a data structure as follows:

```
class Roster {
    List~RosterEntry~ entries
}

class RosterEntry {
    NodeId nodeId
    long weight
    X509Certificate gossipCaCertificate
    PairingPublicKey tssEcPublicKey
    List~ServiceEndpoint~ gossipEndpoints
}

class ServiceEndpoint {
    bytes ipAddressV4
    int32 port
    string domain_name
}

Roster "1" *-- "many" RosterEntry
RosterEntry "1" *-- "many" ServiceEndpoint

```
It is noteworthy that the roster must be immutable to guarantee the integrity of the computed hash.
This map of rosters will typically contain the current Active Roster, an optional previously Active Roster, the optional
previous Candidate Roster (a roster created from a candidate address book), and an optional current Candidate Roster.
There will be new fields in PlatformState - `candidateRosterHash` and `activeRosterHash` - such that at adoption time,
the way to trigger the adoption of a roster will be by the client code inserting a roster in the roster map, alongside
setting the `candidateRosterHash` field in the PlatformState. If a `candidateRosterHash` hash entry already exist in the
map of Rosters, it will be discarded.

> The `activeRosterHash` will be private to the platform and will not be settable or modifiable from outside the platform.

Reviewer 2: Just adding a reminder, that anything in the state is modifiable by the app and nothing is private unless it is "private through obscurity". 

If we removed the `active_roster_hash` then it does become more obscure.   My only concern with removing it is if the semantics on the ordering of the queue change and we start storing more ancient or future rounds of rosters, we might lose track of which is the `active_roster`.

> The `activeRosterHash` will be private to the platform and will not be settable or modifiable from outside the platform.

Author: following yesterday's conversation, the `activeRosterHash` is now removed in favour of the `roundActiveRosters` list.

network's nodes. It will be continuously updated by HAPI transactions, such as those that create, update, or delete
nodes.
The CAB is a snapshot of the Future Address Book (FAB), and it is created when the app decides to adopt a new address
book.

This proposal specifies that Roster(s) will be created from the CAB by the Hedera app and passed to the platform code.
The mechanism for doing so is detailed below.

### Requirements

- Immutability: Rosters must be immutable to protect the integrity of its computed hash.
- Efficient Storage: Roster data should be stored efficiently in the state to minimize overhead and unnecessary data
  copying.
- Clear API: A well-defined API should be provided for the app to create and submit candidate rosters, and for the
  platform to manage the lifecycle of submitted rosters.
- Components update: Platform components that currently rely on the use of the Address book (such as Reconnect, Network,
  Event Validation etc.) must be adapted to use rosters instead.
- Address book paradigm discontinued within the platform codebase, only use rosters instead

### Architecture and/or Components

###### Roster Lifecycle

> ![](TSS%20Roster%20Lifecycle.drawio.svg)

Author: I tried to capture the fact the votes happen as spec'd by the TSS ledger ID proposal but i see the 'No vote' isn't shown to be there. I needed something to specify what happens to an unused Roster and why. I'll see how i can include that within the TSS ledger ID boundary.

> ![](TSS%20Roster%20Lifecycle.drawio.svg)

Author: good catch. Updated. Thanks!

> ![](TSS%20Roster%20Lifecycle.drawio.svg)

Reviewer 2: What I was trying to say is that voting is not relevant in TSS-Roster proposal.  Voting is only relevant in the TSS-Ledger-ID proposal.   Voting happens in response to receiving enough TssMessages and a node says its ready to adopt the keys for the candidate roster.

> ![](TSS%20Roster%20Lifecycle.drawio.svg)

Author: I understand that. However, a roster ends up either being adopted or abandoned. We need to represent this stage in the lifecycle diagram. The 'vote' here is just a label indicating the trigger for either of the end states. I think having that label helps, especially as it's (now) fully enclosed within the TSS ledger ID spec. What if i include a caption "The mechanism of the Yes vs No vote falls outside the scope of this diagram and belongs in the TSS ledger ID spec."?

> ![](TSS%20Roster%20Lifecycle.drawio.svg)

Author: I didn't want to include adoption mechanics in the lifecycle diagram, since it's details for another proposal. I wanted to scope this lifecycle to just the start, transitions, and end state. One thing we can do is change the "Yes vote" label to "Yes vote, after software upgrade". Will that work?

> ![](TSS%20Roster%20Lifecycle.drawio.svg)

Author: Gotcha. Will removing the labels work? Keep the same diagram but without the word `vote` in it? It is important to indicate the end states of rosters in a roster lifecycle diagram, whether that end state is being implemented in this design or otherwise. A lifecycle diagram can involve a journey of a thing from point A to point D, but a single design proposal can focus on the mechanics from A to B, another proposal from B to C and so on.

| Designers          | Kore, Anthony, Ed, Cody, Austin               |
| Functional Impacts | TSS, Genesis, Cryptography, Reconnect         |
| Related Proposals  | TSS-Library, TSS-Ledger-Id, TSS-Block-Signing |
| HIPS               | N/A                                           |

## Purpose and Context

The introduction of the Threshold Signature Scheme (TSS) requires a new mechanism for managing node participation in
consensus and block signing.
The Roster, an immutable subset of the address book, will provide this mechanism, ensuring efficient and secure key
management for TSS operations.
This proposal provides a specification for the behavior of rosters starting from their creation from the
Candidate Address Book (CAB) to their terminal state within the TSS specification.
A roster reaches a terminal state when it is either adopted by the platform or replaced by a new roster.

The Future Address Book (FAB) is a dynamic list maintained within the App that reflects the desired future state of the
network's nodes. It will be continuously updated by HAPI transactions, such as those that create, update, or delete
nodes.

> The CAB is a snapshot of the Future Address Book (FAB), and it is created when the app decides to adopt a new address

Reviewer 7: It is unlikely that something like a Candidate Address Book or Future Address Book will ever exist. Please refrain from using these terms because they are misleading.

Author: Thanks @netopyr. Reworded & updated diagram as discussed offline.

book.

This proposal specifies that a Roster will be created from the CAB by the Hedera app and passed to the
platform code.
The mechanism for doing so is detailed below.

### Requirements

- Immutability: Rosters must be immutable to protect the integrity of its computed hash.
- Efficient Storage: Roster data should be stored efficiently in the state to minimize overhead and unnecessary data
  copying.
- Clear API: A well-defined API should be provided for the app to create and submit candidate rosters, and for the
  platform to manage the lifecycle of submitted rosters.
- Components update: Platform components that currently rely on the use of the Address book (such as Reconnect, Network,
  Event Validation etc.) must be adapted to use rosters instead.
    -
- Address book paradigm discontinued within the platform codebase, only use rosters instead

### Architecture and/or Components

###### Roster Lifecycle

![](TSS%20Roster%20Lifecycle.drawio.svg)

###### Roster Public API

A new method will be added to the platform API to allow the App to submit a Candidate Roster to the platform:

```java

> //in SwirldState

Reviewer 7: It should be noted that this is just a temporary solution as the Inversion of Control effort will most likely change this.

Author: Yes, that's the goal. Services will be responsible for the mechanics of how the roster is created. Platform just need a way to obtain it.

> void setCandidateRoster(@NonNull final Roster candidateRoster);

Reviewer 7: We should document if there are any limitations regarding the thread on which this method can be called (if there are any).

Author: From Platform code's perspective, setting a candidate roster, validating and storing it is expected to be single threaded, with that thread being owned by the Services code that initiates the operation. If i get your question right, I'm not aware of any limitation, maybe other than the single threadedness of it - rosters can be set only one at a time.

```
The Hedera app is already responsible for managing the address book.
We propose that it continues to do so, and when it receives a HAPI transaction, creates a candidate Roster object from
the CAB and set it in the state via the new API.
The State will contain one map of rosters, keyed by roster’s hash - ```Map<Hash<Roster>, Roster>```.

A Roster has a data structure as follows:

```
class Roster {

>     List~RosterEntry~ entries

Reviewer 4: Seems like this should be a map of entries keyed by NodeId rather than a list for easy lookups.

3. All RosterEntry/ies must have a valid X509Certificate.
4. All RosterEntry/ies must have a valid PairingPublicKey.
5. All RosterEntry/ies must have at least one gossip Endpoint.
6. The RosterEntry/ies must be specified in order of ascending node id.
7. All ServiceEndpoint/s must have a valid IP address or domain name (mutually exclusive), and port.
8. The roster must have a unique NodeId for each RosterEntry.

Note that a roster can be valid, but not accepted by the platform.
For example, if a new candidate roster is set via the API, but its hash evaluates to the hash of the existing
candidate roster, the new roster will be discarded. The operation has no effect.

### Core Behaviors, in summary

- Roster Creation: App will create a Candidate Roster from the Candidate Address Book (CAB).
- Roster Submission: App will trigger roster submission by setting the `candidate_roster_hash` field in the
  PlatformState.
- Roster Adoption: The platform will vote to adopt the last submitted Candidate Roster when it is ready, which will be
  on software upgrade.
- Roster Replacement: If a new candidate roster is submitted before the previous one is adopted, the corresponding new

>   Candidate Roster will replace the old one.

Reviewer 7: How does the App learn about the new Roster?

Author: The app will create a new Roster from the address book whenever it decides it's time to start working on adopting a new Roster.

Reviewer 2: @netopyr   The roster replaces the AddressBook in the `ConsensusRound` when it is handled.  It can detect that it is a new consensus roster when it gets back the candidate roster it set.  https://github.com/hashgraph/hedera-services/blob/develop/platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/internal/ConsensusRound.java

If it would make life easier, we could find a more direct way of communication of the change.

Reviewer 7: Does this mean that `ConsensusRound` will become part of the public API?

I like the idea of getting the full roster because we do not need to keep track of potentially multiple candidate rosters. In my opinion, we do not need a more direct way to track the changes. We can copy all values into the address book if we detect a new active roster.

How can we quickly determine if the active roster changed? Comparing each field for each node in each new round would be wasteful.

Reviewer 2: The public API state change that reflects adoption of the candidate roster is in the `RosterState` singleton.    The value of the candidate roster hash will move to the  active roster position, at index 0 in the `round_id_to_roster_pair`.   The round value in the pair indicates the round that the roster became active. 

```
message RosterState {
    byte candidate_roster_hash = 1;
    repeated RoundRosterPair round_id_to_roster_pair = 2;
}
```

     * <p>
     * This value SHALL be the DER encoding of the presented elliptic curve
     * public key.<br/>
     * This field is OPTIONAL (that is, it can initially be null)
     * but once set, it can no longer be null.
     */
    bytes tss_encryption_key = 4;

    /**
     * A list of service endpoints for gossip.
     * <p>
     * These endpoints SHALL represent the published endpoints to which other
     * consensus nodes may _gossip_ transactions.<br/>
     * If the network configuration value `gossipFqdnRestricted` is set, then
     * all endpoints in this list SHALL supply only IP address.<br/>
     * If the network configuration value `gossipFqdnRestricted` is _not_ set,
     * then endpoints in this list MAY supply either IP address or FQDN, but
     * SHALL NOT supply both values for the same endpoint.<br/>
     * This list SHALL NOT be empty.<br/>
     * This list SHALL NOT contain more than `10` entries.<br/>
     * The first two entries in this list SHALL be the endpoints published to
     * all consensus nodes.<br/>
     * All other entries SHALL be reserved for future use.
     */
    repeated proto.ServiceEndpoint gossip_endpoint = 5;
}
```

> The map of rosters will be stored in the state as a PBJ map of Key `hash` and value `Roster`.

Reviewer: I guess we've discussed this offline and arrived to a conclusion that the RosterMap would use the State API maps, while a collection of RosterEntries could indeed be stored in a PBJ map.

#### New Transplant Procedure

The app will decide a network transplant sequence based on the following heuristics:

* Network Transplant process == Active Roster AND State provided
  When both a roster and a state are provided, it signifies a network transplant.
  The node will discard any existing candidate roster, rotate the active roster to the previous roster,
  and adopt the provided roster as the new active roster. This way, the provided network state and roster are adopted.

* Genesis Network process == Active Roster, No State provided
  When a roster is provided, but no pre-existing state exists, it will indicate the creation of a new network.
  The node will initiate the TSS key generation process (out of scope for this proposal),
  create a genesis state with the provided roster, and start participating in consensus from round 0.
* Keep Network Settings (Normal restart) == No Active Roster AND State provided
  If only a state is provided without a roster, the node will retain its existing network settings, including the active
  roster.
  This will be the typical behavior for a node restarting within an established network.

>   ![](Proposed%20Network%20Transplant%20flow.drawio.svg)

Author: Answering your second question first, the State is not always required. For example, the case of genesis network requires an active roster but no state.

For the first, I'm open to calling it override roster if it helps. @edward-swirldslabs, thoughts?

>   ![](Proposed%20Network%20Transplant%20flow.drawio.svg)

Author: Thanks for the feedback. We have renamed this to 'Genesis Roster' with an accompanying description.

>   ![](Proposed%20Network%20Transplant%20flow.drawio.svg)

Reviewer: Regarding the first question, my point is exactly to make the state a required input argument. Due to the IOC, we can delegate the creation of the genesis state to the app completely. What do you guys think about this?

>   ![](Proposed%20Network%20Transplant%20flow.drawio.svg)

Reviewer 2: It depends on whether or not there are genesis fields we need to initialize in the state for the platform's purposes.

>   ![](Proposed%20Network%20Transplant%20flow.drawio.svg)

Author: A method will need to be added to PlatformBuilder for genesis as follows:
`void withGenesisRoster(@NonNull final Roster genesisRoster)`
which will be responsible for setting the GenesisRoster when needed. The app can conditionally call this method with a Roster if it determines the startup sequence requires it, though the mechanics of that is upon Services to implement.

>   ![](Proposed%20Network%20Transplant%20flow.drawio.svg)

Reviewer: To reiterate, I specifically question the need for this extra method/extra parameter. I believe we can nuke the genesis roster thing. I believe the app can install a genesis roster directly into the state before passing it to the platform builder on startup. This would simplify our interface with the app.

Would anything break if we do so?

>   ![](Proposed%20Network%20Transplant%20flow.drawio.svg)

Reviewer: Per comments in @edward-swirldslabs 's TSS-Ledger-Id design proposal, it seems that we want to encapsulate roster handling on the Platform side, and from this perspective, accepting a genesis roster as an argument kind of makes sense. So I guess this conversation can be resolved for now.

## Core Behaviors, in summary

- Roster Creation: Hedera App will create a Candidate Roster from the Address Book.
- Roster Submission: App will trigger roster submission by setting the `Roster` object in the
  `PlatformState` API.
- Roster Validation: The submitted roster is validated and hashed, if valid, and the hash is not already in the state.
- Roster Storage: The Roster is stored in the State as a map of Roster Hash to Roster. A reference to the candidate
  roster's hash is also stored in the
  `RosterState` object.

> - Roster Adoption: The adoption of the candidate roster is beyond the scope of this proposal.

Reviewer: I'm unsure if it's fair because this makes this design proposal incomplete. If/when this design is implemented and launched, there's no way to actually switch to the candidate roster. This makes this design depend on another project that must be launched together with this one in order to provide at least some adoption mechanism - for example, and unconditional adoption on a software upgrade.

Author: That is an acceptable dependency from project planning perspective. The artifacts of this proposal form the building blocks to the follow-up proposal - TSS Ledger ID. When TSS Rosters is completed, the candidate Roster will just sit in the state until the work to adopt it is completed. That helps avoid a big-bang release.

Author: Thanks for raising this. An important clarification here is that candidate roster is always adopted on the next software upgrade.

Reviewer: Thanks for clarifying!

     * but a node SHALL NOT, ever, serve multiple shards or realms,
     * therefore the node identifier MAY be repeated _between_ shards and realms.
     */
    uint64 node_id = 1;

    /**
     * A consensus weight.
     * <p>
     * Each node SHALL have a weight of zero or more in consensus calculations.<br/>
     * The consensus weight of a node SHALL be calculated based on the amount
     * of HBAR staked to that node.<br/>
     * Consensus SHALL be calculated based on agreement of greater than `2/3`
     * of the total `weight` value of all nodes on the network.
     */
    uint64 weight = 2;

    /**
     * An RSA public certificate used for signing gossip events.
     * <p>
     * This value SHALL be a certificate of a type permitted for gossip
     * signatures.<br/>
     * This value SHALL be the DER encoding of the certificate presented.<br/>
     * This field is REQUIRED and MUST NOT be empty.
     */
    bytes gossip_ca_certificate = 3;

    /**
     * An elliptic curve public encryption key.<br/>
     * This contains the _long term_ public key for this node.
     * <p>
     * This value SHALL be the DER encoding of the presented elliptic curve
     * public key.<br/>

>      * This field is REQUIRED and MUST NOT be empty.

Author: updated the doc.

     * If the network configuration value `gossipFqdnRestricted` is set, then
     * all endpoints in this list SHALL supply only IP address.<br/>
     * If the network configuration value `gossipFqdnRestricted` is _not_ set,
     * then endpoints in this list MAY supply either IP address or FQDN, but
     * SHALL NOT supply both values for the same endpoint.<br/>
     * This list SHALL NOT be empty.<br/>
     * This list SHALL NOT contain more than `10` entries.<br/>
     * The first two entries in this list SHALL be the endpoints published to
     * all consensus nodes.<br/>
     * All other entries SHALL be reserved for future use.
     */
    repeated proto.ServiceEndpoint gossip_endpoint = 5;
}
```

The map of rosters will be stored in the state as a virtual map of Key `Hash<Roster>` and value `Roster`.
The `Hash` object(SHA_384), which is the key of the virtual map, will reuse an existing protobuf that looks as follows:
```proto
/**
 * List of hash algorithms
 */
enum HashAlgorithm {
  HASH_ALGORITHM_UNKNOWN = 0;
  SHA_384 = 1;
}

/**
 * Encapsulates an object hash so that additional hash algorithms
 * can be added in the future without requiring a breaking change.
 */

> message HashObject {

Author: It's intended to be the key of the virtual map, so it must be a protobuf object. And since the Hash protobuf object already exist, we reuse it. I've only included it here for completeness.

> message HashObject {

Reviewer 2: Where does it already exist?   Can you show me pre-existing usage?  I'd like to learn.

> message HashObject {

Reviewer 5: We shouldn't reuse the "streams" HashObject and HashAlgorithm.
Both are not well structured (something Jasper brought up in a different context).
The Block Stream is adding a better defined HashAlgorithm to basic_types.proto:
```protobuf
/**
 * A specific hash algorithm.
 *
 * We did not reuse Record Stream `HashAlgorithm` here because in all cases,
 * currently, this will be `SHA_384` and if that is the default value then
 * we can save space by not serializing it, whereas `HASH_ALGORITHM_UNKNOWN`
 * is the default for Record Stream `HashAlgorithm`.
 */
enum BlockHashAlgorithm {
    /**
     * A SHA2 algorithm SHA-384 hash.
     * <p>
     * This is the default value, if a field of this enumerated type is
     * not set, then this is the value that will be decoded when the
     * serialized message is read.
     */
    SHA_384 = 0;
}
```

Regarding `HashObject`, it might be better to just use a `bytes` for the hash value directly, and specify the hash algorithm, where needed (which I believe is somewhat rare), on the message that contains the hash value.

> message HashObject {

Author: thanks @jsync-swirlds for the context. @edward-swirldslabs I considered the HashObject as I noticed it's used in a bunch of utils(`RecordSigningUtils` and `FileSignTool` are examples). Additionally, the initial map implementation was supposed to be PBJ Map, which does not allow byte-type keys. 

We have however since concluded on using `VirtualMap` instead. So, to use a `byte` key as suggested, we will have to provide an implementation of `VirtualKey` or similar. I've removed the HashObject and replaced with this explanation.

> message HashObject {

Reviewer 2: @kfa-aguda, when you say that the VirtualMap doesn't allow bytes for keys, are you saying this is not possible as a key? 

```
message WrappedKey{
   bytes byteArray;
}

```

> message HashObject {

Reviewer 2: I looked it up, only integers and strings are allowed for keys.  complex objects are not allowed as keys in pbj maps.    This issue is a non-issue since we're not using pbj maps here.

> message HashObject {

Author: This probably works the PBJ map. I didn't implement it as subsequent conversations shifted to using `VirtualMap`. Now, we're back to using PBJ map, so I've consulted @imalygin for recommendation. I think a simple byte wrapper like you proposed should be sufficient.

> message HashObject {

Reviewer 5: If we're wrapping bytes, it may be appropriate to use the "well known type" `google.protobuf.BytesValue` rather than creating our own message for the same purpose.

> message HashObject {

Author: agreed on using a well known type, although @ivan suggested `com.hedera.pbj.runtime.io.buffer.Bytes`. Happy to work that out during implementation.

  int32 length = 2;

  /**
   * Specifies the result of the hashing operation in bytes
   */
  bytes hash = 3;
}
```

The `Roster` value is modeled as previously shown.

The platform state protobuf will be updated to include the roster hashes.

```proto
/**
 * The current state of platform consensus.<br/>
 * This message stores the current consensus data for the platform
 * in network state.
 *
 * The platform state SHALL represent the latest round's consensus.<br/>
 * This data SHALL be used to ensure consistency and provide critical data for
 * restart and reconnect.
 */
message PlatformState {

    //... existing entries omitted for brevity ...

    /**
     * The SHA-384 hash of a candidate roster.
     * <p>
     * This is the hash of the roster that is currently being considered
     * for adoption.
     * A Node SHALL NOT, ever, have more than one candidate roster
     * at the same time.
     */
    bytes candidate_roster_hash = 6;

    /**
     * The SHA-384 hash of an active roster
     * This is the hash of the roster that has already been adopted by the
     * network.
     * A Node SHALL NOT, ever, have more than one active roster
     * at the same time.
     */

>     bytes active_roster_hash = 7;

Reviewer 2: this comment is superceded by a discussion that happened on July 22

}
```

It is noteworthy that the roster must be immutable to guarantee the integrity of the computed hash.
This map of rosters will typically contain the current Active Roster, an optional previously Active Roster, and an
optional current Candidate Roster.
There will be new fields in PlatformState - `candidateRosterHash` and `activeRosterHash` - such that at adoption time,
the app will set the roster through the API and the platform code inserting this roster into the roster map, alongside
setting the `candidateRosterHash` field in the PlatformState. If a `candidateRosterHash` hash entry already exist in the

> map of Rosters, it will be discarded.

Author: >If a `candidateRosterHash` hash entry already exist in the
map of Rosters, it will be discarded.
I think that describes idempotency, but I've included the word specifically. Is that enough?

> map of Rosters, it will be discarded.

Reviewer: Looks good now. Thanks!

The `activeRosterHash` will be private to the platform and will not be settable or modifiable from outside the platform.

This indirection (using a Map instead of a Singleton) avoids moving data around in the merkle state which requires data
to be copied.
When a map is updated, only the key-value pair that changes needs to be added to the block stream.
Another benefit of this approach is that adoption trigger becomes simple (app sets the roster) with delineated
responsibilities between the app and the platform.

This map will not grow infinitely. Insertion of rosters will be controlled by ensuring that the acceptance of a new
Candidate Roster

> invalidates and removes the current candidate roster.

Author: I've moved this part up into where we describe the map.

###### Roster Validity

In simple terms, the following constitutes a valid roster:

1. The roster must have at least one RosterEntry.
2. At least one RosterEntry/ies must have a non-zero weight.
3. All RosterEntry/ies must have a valid X509Certificate.
4. All RosterEntry/ies must have a valid PairingPublicKey.
5. All RosterEntry/ies must have at least one gossip Endpoint.
6. The RosterEntry/ies must be specified in order of ascending node id.
7. All ServiceEndpoint/s must have a valid IP address or domain name (mutually exclusive), and port.
8. The roster must have a unique NodeId for each RosterEntry.

Note that a roster can be valid, but not accepted by the platform.
For example, if a new candidate roster is set via the API, but its hash evaluates to the hash of the existing

> candidate roster, the new roster will be discarded. The operation has no effect.

Author: Here, we are trying to clarify why a roster can be valid but not accepted, with an example. I think this part is necessary, and must be separate from the behavior of the map.

> candidate roster, the new roster will be discarded. The operation has no effect.

Reviewer: Sounds good.

### Core Behaviors, in summary

- Roster Creation: App will create a Candidate Roster from the Candidate Address Book (CAB).

> - Roster Submission: App will trigger roster submission by setting the `candidateRosterHash` field in the PlatformState.

Reviewer 2: ```suggestion
- Roster Submission: App will trigger roster submission by setting the `candidate_roster_hash` field in the PlatformState.
```

- Roster Adoption: The platform will vote to adopt the last submitted Candidate Roster when it is ready, based on

>   specifications outlined in the TSS Ledger ID Proposal (referenced under related Proposals).

Reviewer 2: For this proposal, it is adopted at software upgrade

}
```

### Roster changes needed for Components

#### Block Proof

In a fully Dynamic Address Book (DAB) where, say, the roster for round 101 might be different from the roster for round
102, there is a need to track what roster is used for each non-ancient round.
We propose the introduction of a queue of roster hashes stored in the state, where the index of a roster hash determines
what round the roster is active for as shown.
![](TSS%20Roster%20Lifecycle-Roster%20and%20Rounds.drawio.svg)
The size of this queue would be ancient round window + roster offset.
The roster offset refers to the number of rounds in advance that the network determines and declares the active roster
for a specific future round.
This acts as a proactive mechanism that allows the network to prepare for upcoming changes in the roster and ensure a
valid transition to a future roster.
When each round is handled, we will pop off the active roster hash for the round that just became ancient, and add a new
active roster hash for a new future round.

Before full DAB (DAB Phase 2.5), we propose to store the current and previous active rosters in this newly introduced
queue of hashes.

> In the DAB Phase 2.5 implementation, this queue will contain only 2 hashes, with the first hash representing

Reviewer 2: The term `DAB Phase 2.5` is not referencing a formal deliverable.  It's an informal term I've used to refer to the changes that we're incorporating.  It may confuse external reviewers (PM, Services, DevOps, etc) that are not onboarded with the meaning.   It might be better to simply remove the term and say `In this proposal implementation ...`

Author: updated, thanks!

the previous active roster, and the second hash representing the current active roster.
At upgrade boundaries, we will pop off the previous active roster, and add a new roster to the end of the queue, which
is the new active roster.

This approach provides the benefit of introducing the necessary data structure for full DAB, while also providing a

> mechanism for tracking the active roster for each round in the interim.

Reviewer: If this queue exists, then it already provides a reference to the active roster, making a pointer to the active roster in the platform state as described above unnecessary. Should we remove it from the platform state and only store a pointer to the candidate roster (or null if none) in there?

Author: I am working on this bit, it's a little fuzzy at the moment. I'll clarify today.

Author: Actually, this is different. The map is Map<Hashes,Roster>, which is stored in the State. The map is unaware of which roster is active or otherwise. The Roster itself does not carry this information either. We still need a mechanism to track which of the stored rosters is active. That mechanism is what is addressed by introducing a queue of 2 items (roster hashes) whose tail is the active roster hash. So, to get the active roster, we retrieve the tail of the queue, and look it up in the map.

Reviewer: Right. I'm not talking about the map - the map should remain as it is described. I'm talking about the HashObject pointers in the platform state. While a pointer to the candidate roster should be there, the pointers to the active and previous rosters seem redundant because they'll be present in this queue as well.

Author: As per the architecture review last friday, this has been removed. We now keep a pointer to only the candidate roster's hash value.

Reviewer: Sounds good. Thanks!

  state. Both of these will need to be updated to use rosters instead.

#### Others
- Networking. Most parts of the network code have been abstracted away from the address book as part of the Dynamic
  Address Book effort, so there should be minimal work left there.
- Event Validation. The event validation logic uses the address book to determine whether a given event has a valid
  signature. This will be updated to use rosters instead.
- Miscellaneous components. Other components that reference things stored in the address book (like consensus needing
  the node weight or the crypto module verifying keys) will need to be updated to use things stored in rosters instead.

### Test Plan

Some of the obvious test cases to be covered in the plan include validating one or more of the following scenarios:

1. New valid Candidate Roster created with no subsequent one sent by app. Verify accept.
2. New valid Candidate Roster created with subsequent one sent by app. Verify accept.
3. Invalid roster(s) sent by the app. Verify reject.
4. Node Failures During Roster Change: What happens if nodes fails or disconnects during a roster change? Verify valid

>    node successfully reconnects.

Reviewer: Actually, this one is interesting. Suppose a node is dead during a software upgrade when the rest of the network adopts a new roster.

Once the dead node resumes after the network has adopted a new roster, will it be able to rejoin the network? At this point, this node has a pointer to the network's current active roster in its candidate roster pointer.

To complicate this a bit further, suppose the node has been dead during two consecutive software upgrades and the network has replaced the roster during both, so the dead node has absolutely no references to the current roster at all, even as a candidate roster. And the rest of the network has already forgotten about the "previous roster" that was used when this node was last alive. Will it still be able to reconnect?

Do we even support such scenarios or we require all nodes to be alive during every software upgrade/roster adoption event? If so, I think we should document this assumption to ensure we don't miss any edge cases that we need to support.

Author: The node will be able to rejoin the network because it's key is in the active roster. If it's able to join the network and reconnect, it will get the latest state which will contain the active roster. @edward-swirdlabs, is that correct?

Reviewer 2: Are nodes allowed to reconnect with each other across software upgrade boundaries?   Or do nodes have to be on the same software version to perform a reconnect?  

Suppose a node is down for a long time and doesn't receive the freeze-command.  The rest of the network performs a software upgrade.  The node is restarted and comes back up.  What happens?  @litt3 may know off the top of his head.  Otherwise we'll need to look at the code. 

I don't expect anything to change about reconnect with the TSS-Roster proposal.    We might need to speak with services about what would happen in response to DAB Phase 2.  Our behavior will be inherited from where they leave off.

Reviewer: If I understand correctly, a node operator would have to upgrade the software (apparently, upload latest .jars or similar) to the node.

Assuming this is correct, and assuming that the node still has a state on disk saved with the previous version of the software, or, in the second scenario, with two-versions-old software, would it be able to rejoin the network then?

The question is more about whether we support such scenarios officially rather than whether the current code is capable of doing this. If we do support them, then we need tests for these scenarios. If we don't support them, then let's explicitly state so in the testing plan.

Author: @anthony-swirldslabs - i think this is a valid question. Mind asking in one of the channels?

Reviewer: Just posted a question to the engineering-team channel.

Reviewer: @kfa-aguda - so, you can find all the details in that slack channel. But long story short, both the scenarios above are supported.

To summarize and rephrase, a node X can be dead for as long as it pleases, even for years, hence skipping multiple roster updates. It should be able to rejoin the network w/o any manual interventions as long as both the following conditions are true:
1. The node X itself remains in the roster during all this time. In fact, I think it may be removed in the middle of that period, but the main point is that it was present in the roster when it died, and it is present in the roster when it comes back alive.
2. The old version of the roster on node X that was preserved when it died should contain at least one node (e.g. node Y) that is still active on the network when the node X comes back alive. It is so that the node X can connect to at least that node Y in order to start a reconnect session and catch up with the network.

While the above is mostly handled by the reconnect itself, the Roster proposal introduces a new implication because it preserves the references to the active and candidate (and potentially, one or many past rosters) in the state. So we need to test at least two scenarios explicitly as described in the original comment:
1. The node X skips a single upgrade and goes through it after the rest of the network has upgraded and updated the roster.
2. The node X skips several upgrades/several roster changes, and comes back online at a later time. As long as the conditions described above are satisfied, the node should still be able to rejoin and update its roster to the latest one.

Please include these test scenarios in the testing plan.

Author: Thanks for bringing the update here @anthony-swirldslabs. Just saw the conversation. I think (for completeness purposes) there is one more condition discussed - that the node X runs the same software version as the rest of the network. I'll add these test cases as described.

# TSS - Roster Proposal

---

## Summary

This proposal outlines the design specification of rosters to support Threshold Signature Scheme (TSS) block signing.

> It details the roster's lifecycle, data storage, API, DevOps workflow, and changes necessary changes to associated

Reviewer 8: ```suggestion
It details the roster's lifecycle, data storage, API, DevOps workflow, and necessary changes to associated
```

| Related Proposals  | TSS-Library, TSS-Ledger-Id, TSS-Block-Signing |
| HIPS               | N/A                                           |

## Purpose and Context

The introduction of the Threshold Signature Scheme (TSS) requires a new mechanism for managing node participation in
consensus and block signing.
The Roster, a subset of the address book, will provide the data structure and API for this mechanism.
This proposal provides a specification for the behavior of rosters starting from their creation from the Address Book to
their terminal state.
A roster reaches a terminal state when it is either adopted by the platform or replaced by a new roster.
The candidate roster will always get adopted on the next software upgrade.

The Hedera App (henceforth reffered to as 'App') will maintain a version of the Address Book as a dynamic list that
reflects the desired future state of
the
network's nodes. This dynamic list of Addresses will be continuously updated by HAPI transactions, such as those that
create, update, or delete nodes.
At some point when the App decides it's time to begin work on adopting a new address book (scope beyond this proposal),

> it will create a `Roster` object with information from the Address Book and pass it to the platform.

Reviewer 8: Style note: Can we get newlines between paragraphs? Or if these are part of one paragraph, remove the newlines at the end of each sentence?

Author: I'll make the paragraph changes. For removing newlines, our code style has a max line length I tried to adapt to. Perhaps we can make an exception for design docs?

Reviewer 8: Max line-length is good. But clearly there is something else going on, otherwise you wouldn't have lines like 29, 30, and 31.

The mechanism for doing so is detailed below.

## What this proposal does not address

> 

Author: Deleting in the next push.

## Requirements

- Immutability: The roster hash will be a critical piece of data. Rosters must be immutable to protect the integrity
  of its computed hash.
- Efficient Storage: Roster data should be stored efficiently in the state to minimize overhead and unnecessary data
  copying.
- Clear API: A well-defined API should be provided for the App to create and submit candidate rosters to the platform,
  and for the
  platform to manage the lifecycle of submitted rosters.
- Components update: Platform components that currently rely on the use of the Address book (such as Reconnect, Network,
  Event Validation etc.) must be adapted to use the new `Roster` object instead.
- Address book paradigm discontinued within the platform codebase in favor of `Roster`s

## Architecture

###### Roster Lifecycle

![](TSS%20Roster%20Lifecycle.drawio.svg)

### Roster Public API

#### Setting the Candidate Roster
A new method will be added to the platform state to allow the App to submit a Candidate Roster to the platform:

```java

> //in PlaformState

Reviewer 4: This is no longer accurate. The roster will be added to the RosterMap and the hash of the candidate roster will be stored in the RosterState singleton https://app.diagrams.net/#G1OD12O2KqycNWIzKu6vP-ExPEGKH9-8De#%7B%22pageId%22%3A%22KP1hL-13WYacigEbZWgY%22%7D

Author: Current approach is to have Services call this method with a Roster. Once done, subsequent routines are called to validate the Roster, compute the hash and persist both the Roster and hash in the various data structures in the durable state. This is still valid. We need to be able to reject the Roster synchronously if invalid.

void setCandidateRoster(@NonNull final Roster candidateRoster);
```

The steps to validate and add the Roster to the state will be executed within the same thread that submits the request.
An exception will be thrown if the Roster is invalid.
The durability of the submitted Roster is ensured via storing it in the state.

The existing `PlatformState` class is in the middle of a refactor, and in the long term, may cease to exist.

> However, its replacement will continue to have this new API to set a candidate roster.

Reviewer 4: In the project tree breakdown, we said that part of this proposal is to create the RosterSingleton state and the RosterMap state

https://app.diagrams.net/#G10RSm15R0z_Fu4_zwTSQOr--r4DZPBT4B#%7B%22pageId%22%3A%22sQsfR1DOcLvvdrg7Kwju%22%7D

However, its replacement will continue to have this new API to set a candidate roster.

## Data Structure

The App is already responsible for managing the address book.
We propose that it continues to do so, and when it receives a HAPI transaction, creates a candidate Roster object

> and passes it to the platform via the new API. The platform will then validate the roster and store it in the state.

Reviewer 8: In reality a new candidate roster will be generated at most once per day (first transaction after midnight). In the end, it is up to the app to decide when to propose a new roster.

### Protobuf

#### Roster

The protobuf definition for the Roster will look as follows:

```proto
/**
 * A single roster in the network state.
 * <p>
 * The roster SHALL be a map of key `NodeId`, value `RosterEntry`.
 */
message Roster {

    /**
     * Map of roster entries, one per consensus node.
     * <p>

>      * This map SHALL contain roster entries in natural order of ascending node ids.

Reviewer 8: I think I misunderstood this statement. I thought it was saying this was an ordered map, but what it is saying is that it will have node ids (and incidentally, node Ids are ascending). I'm not sure what value is added by "in natural order of ascending node ids" and might be clearer if this were not stated.

     * The node Id key SHALL match the node Id in the value.
     * This map SHALL NOT be empty.<br/>
     */

>     map<uint64, RosterEntry> rosterEntries = 1;

Reviewer 9: I would suggest
```
repeated RosterEntry roster_entries = 1;
```
because serialization will be faster, the serialized size will be smaller, and the overhead of building a `Map<Long, RosterEntry>` from the list will be vanishingly small.

Author: The original proposal was a list but changed to a map after a feedback that we do need the roster entries in a form where their keys are sorted.

If i understand you correctly, you're saying we should still build and hold a map in memory, but store a list (of the map's entries) in the state?

Reviewer: While I don't have a strong opinion on this, I'll just note that a nice property of the protobuf map is the deterministic serialization of the map object. With a list, the serialized bytes will depend on the order of items in the list, making the user of the model responsible for deterministic binary representation. With the map, the user doesn't need to worry about that and they always get the deterministic result - all entries are always serialized in the natural order of the keys.

The determinism of the binary form might be important for hashing/signing and similar use-cases.

Reviewer 2: If our intent is to use the protobuf as in-memory storage instead of copying the data into a Java Map on deserialization, then lookup on a list is O(n), but O(1) in a hash table and O(logn) in a tree.  I don't know what the backing implementation is for the State API, but it has to be better than lookup in a list.

Reviewer 7: It probably does not matter here, but the asymptotic growth rate (big o notation) is irrelevant for small datasets. Lookups in small ArrayLists are usually faster than in HashMaps.

Reviewer 8: https://protobuf.dev/programming-guides/proto3/#maps seems to say that maps are dangerous from a deterministic wire format perspective?

> Wire format ordering and map iteration ordering of map values are undefined, so you cannot rely on your map items being in a particular order.

Reviewer: In PBJ, we made map serialization deterministic though. We always serialize them in the natural order of keys. And upon deserialization, we always store them in PbjMap objects which expose the order via their `public List<K> getSortedKeys()` method. I should've clarified that in my comment above instead of calling them "protobuf maps".

So unless we want to stop using PBJ, this shouldn't be an issue.

Reviewer 5: It may be important to note that these values (and all values in state) will be part of the _public API_ in the Block Stream, and may be processed by protoC-generated code (or other protobuf compilers for languages, like Rust, that protoC doesn't support).

Those other systems may _not_ be deterministic for map ordering (possibly making it very difficult for those systems to correctly recreate state).

Just something to consider.

Reviewer 8: We decided to go with the recommendation from Michael and forbid the use of maps in state. While it would be awesome to have it be a map, so that the generated pbj classes would give us a map that behaved just as we liked, the fact is that map on-wire serialization is non-deterministic across multiple protoc implementations, and since state is intended to be parsed and used by downstream applications across multiple languages, using map will create friction for those people.

}
```

#### RosterEntry

A roster entry will look as follows:

```proto
/**
 * A single roster entry in the network state.
 *
 * Each roster entry in SHALL encapsulate the elements required
 * to manage node participation in the Threshold Signature Scheme (TSS).<br/>
 * All fields are REQUIRED.
 */
message RosterEntry {

    /**
     * A consensus node identifier.
     * <p>
     * Node identifiers SHALL be unique _within_ a shard and realm,
     * but a node SHALL NOT, ever, serve multiple shards or realms,
     * therefore the node identifier MAY be repeated _between_ shards and realms.
     */

>     uint64 node_id = 1;

Reviewer 8: Node ID's are intended to be _ledger global_, not scoped within a shard and/or realm.

Author: This doc was lifted from the existing protobuf definition of `node_id` so I suspect we have this error everywhere. @jsync-swirlds, you might find this of note.

Reviewer 5: Noted.
We'll modify this the next time we update the node protobuf files. Given we don't have shards or realms and haven't figured out how the counter might be communicated across shards and realms, it doesn't seem worth a separate HIP just for this item, though.

Reviewer 8: Let's please update the language here so it is correct, and then we can fix this in the other protobuf files separately.

Author: updated.

    /**
     * A consensus weight.
     * <p>
     * Each node SHALL have a weight of zero or more in consensus calculations.<br/>
     * The consensus weight of a node SHALL be calculated based on the amount

>      * of HBAR staked to that node.<br/>

Reviewer 8: While it is true that this will be based on the HBAR for mainnet, it isn't necessarily true for all uses of a roster. The roster could be used by a private network, using something other than HBAR. Or it could be a private network that just gives equal weights, or something else random.

The definition shouldn't be tied to HBAR, it should instead just say that the sum of the weights of all nodes in the roster form the total weight of the system, and each node's individual weight is proportional to that sum.

Author: updated.

     * Consensus SHALL be calculated based on agreement of greater than `2/3`
     * of the total `weight` value of all nodes on the network.
     */
    uint64 weight = 2;

    /**
     * An RSA public certificate used for signing gossip events.
     * <p>
     * This value SHALL be a certificate of a type permitted for gossip
     * signatures.<br/>
     * This value SHALL be the DER encoding of the certificate presented.<br/>
     * This field is REQUIRED and MUST NOT be empty.
     */
    bytes gossip_ca_certificate = 3;

    /**
     * An elliptic curve public signing key.<br/>
     * This contains the _long term_ public key for this node.
     * <p>
     * This value SHALL be the DER encoding of the presented elliptic curve

>      * public key.<br/>

Reviewer 8: Should we spec which kinds of key curves are supported?

Author: I'm not sure. This is a byte field, so the supported key curves will not change its invariant if/when they change. We may opt to document this fact for informational purposes, but that should live in the java code.

Reviewer 2: I think we should indicate the curve, just in case someone doesn't trust our CLI tool and wants to generate their `tssEncryptionKey` with another tool.

Reviewer 8: Why is there not an RSA encryption key field? The roster used by the consensus system needs a key field and curve type so the event intake validation can verify the event, but doesn't specifically need to know anything about TSS. On the other hand, the roster used by the execution system also needs to save in state the TSS key (I assume), in which case we need both types of fields here?

     * This field is OPTIONAL (that is, it can initially be null)
     * but once set, it can no longer be null.
     */
    bytes tss_encryption_key = 4;

    /**
     * A list of service endpoints for gossip.
     * <p>
     * These endpoints SHALL represent the published endpoints to which other
     * consensus nodes may _gossip_ transactions.<br/>
     * If the network configuration value `gossipFqdnRestricted` is set, then
     * all endpoints in this list SHALL supply only IP address.<br/>
     * If the network configuration value `gossipFqdnRestricted` is _not_ set,
     * then endpoints in this list MAY supply either IP address or FQDN, but
     * SHALL NOT supply both values for the same endpoint.<br/>
     * This list SHALL NOT be empty.<br/>
     * This list SHALL NOT contain more than `10` entries.<br/>
     * The first two entries in this list SHALL be the endpoints published to

>      * all consensus nodes.<br/>

Reviewer 8: I don't understand this. All nodes will know of all endpoints, since the roster is stored in state?

Author: Grandfathered comment from `AddressBook`'s proto. All of the invariants of the address book are supposed to apply here. @jsync-swirlds, thoughts?

Reviewer 2: I agree with Richard, the roster is in the state, so all entries are visible to all parties.

Author: nuked.

     * All other entries SHALL be reserved for future use.
     */
    repeated proto.ServiceEndpoint gossip_endpoint = 5;
}
```

#### Roster Map

A States API map `Roster Map` (separate from the PBJ Map modeled above) will be introduced.
It will contain keys which will be the hash of the roster (of type `byte`) and values of type `Roster`.
This map will store all candidate and active rosters.

> It will be an implementation of `ReadableKVState<>` and `WriteableKVState<>` interfaces.

Reviewer 9: It's very unlikely this will be a new implementation of these interfaces; it would be an instance of an `OnDiskWritableKVState`.

I think the indirection here makes things a lot less explicit, however. I would suggest instead three `Singleton` states:
1. _(Always set)_ A `RoundRosterPair` with the current active roster and the round it became active.
2. _(Sometimes set)_ A `RoundRosterPair` with the previous active roster and the  round _it_ became active.
3. _(Sometimes set)_ A `Roster` with the candidate roster.

Author: Ah, I've just looked again. Agreed on the instance of OnDiskWritableKVState.

We do need a `Roster` store that is forward-looking, considering we have Dynamic Address Book (DAB) on the horizon. In DAB, there may be many previous active `Roster`s we must keep track of. The use of singletons for storing the `Roster` today means we introduce a lot more state-change/data structure work to get to DAB. I'm not sure that's worth it. @jasperpotts, thoughts?

Reviewer 4: I am torn on this. One one had, i don't think we should limit the number of rosters we can store in state by creating specific singletons for each. In fully dynamic address book, we need to track the active roster for every non-ancient round. It's true that even with fully dynamic address book, it's hard to image we will ever have more than the three rosters in state at one time especially given that it takes time to adopt a candidate roster, and 26 rounds is about 26 seconds today. But if things change in the future, like increasing the number of rounds non-ancient or decreasing the time for adoption of a new roster - we would need to change the state to support more rosters.

On the other hand, we know that we want to _decrease_ the number of rounds non-ancient in the future, the chance of ever needing more than these three is very small, and making them explicit is much clearer.

I do not think that storing them as a list in a singleton is the best option because that will increase the size of the block stream.

     * at the same time.
     */
    byte candidate_roster_hash = 1;

    /**
     * A list of round numbers and roster hashes.
     * The round number indicates the round in which the corresponding roster became active
     * <p>
     * This list SHALL be ordered by round numbers.<br/>
     * This list MUST be deterministically ordered.
     */
    repeated RoundRosterPair round_roster_pairs = 2;
}
```

#### RoundRosterPair

A `RoundRosterPair` will be defined as follows:

```proto
/**
 * A pair of round number and active roster hash.
 * <p>
 * This message SHALL encapsulate the round number and the hash of the
 * active roster used for that round.
 */
message RoundRosterPair {

    /**
     * The round number.
     * <p>
     * This value SHALL be the round number of the consensus round in which this roster became active.
     */
    uint64 round_number = 1;

    /**
     * The SHA-384 hash of the active roster for the given round number.
     * <p>
     * This value SHALL be the hash of the active roster used for the round.
     */
    byte active_roster_hash = 2;
}
```

It is noteworthy that the roster must be immutable to guarantee the integrity of the computed hash.

> This map of rosters will typically contain the current Active Roster, an optional previously Active Roster, and an

Reviewer 8: What map is this? Do you mean the round_roster_pairs **list**?

Reviewer 8: Ah, I think this refers to the "Roster Map" (i.e. virtual map)

 * This message SHALL encapsulate the round number and the hash of the
 * active roster used for that round.
 */
message RoundRosterPair {

    /**
     * The round number.
     * <p>
     * This value SHALL be the round number of the consensus round.
     */
    uint64 round_number = 1;

    /**
     * The SHA-384 hash of the active roster for the given round number.
     * <p>
     * This value SHALL be the hash of the active roster used for the round.
     */
    byte active_roster_hash = 2;
}
```
It is noteworthy that the roster must be immutable to guarantee the integrity of the computed hash.
This map of rosters will typically contain the current Active Roster, an optional previously Active Roster, and an

> optional current Candidate Roster. Insertion of rosters will be controlled by ensuring that the acceptance of a new

Reviewer 4: Doesn't the candidate roster goes in the RosterMap, not the RosterState? If i'm wrong, then the comment for `active_roster_hash` and the field name must change.

> responsibilities between the App and the platform.

Reviewer 8: I don't understand this design. We have a virtual map "Roster Map" with hashOfRoster->Roster. And we have RosterState (which is a singleton?) with the hash of the current roster and then a list of round/hashOfRoster as to what round the roster became active in.

First, what is the advantage of the RosterMap? Why not just have RosterState contain the actual candidate roster and list of round/roster? Why have it be hashes?

Second, if there are only ever going to be 3 items in the RosterMap, why use a map and not just a singleton with 3 fields?

Third, the management of the list of round_roster_pairs sounds complex. What are the theoretical minimum and maximum number of entries in this list?

Author: Answering the first question last:

2. `RosterMap` is the durable storage for all rosters that must be stored in the state, regardless of their type. It is expected to be a few items, yes, but not necessarily 3. In theory, in the full DAB case, we could have multiple previous active rosters between all the current non-ancient rounds (4? 5? 6?). Using a singleton with n fields means the entire thing has to be re-hashed each time (polluting the record stream?). So we opted for a state data structure that avoids that, a map.

3. Before DAB, the round_roster_pairs list will have two items. The item at index 0 will be the active roster hash and the round it became active, index 1 will be for the previous active roster. Most times, the list will be of size 1. The period immediately following an upgrade, it will have 2 items. After DAB, the theoretical size is much higher than 2, but realistically less than say, 10. The consideration here is that since we know index 0 will always be the active roster, it’s cheaper to just get element at that known index. We could replace all of these with singletons, but it takes us farther away from DAB implementation, not closer.

Now, back to question 1
1. The existence of the round number in the `RosterState` makes it slightly faster changing than the RosterMap, which means it will need more rehashing. If this is correct, and we move the Roster into it, we will have to re-hash the rosters many more times than the current design requires. The Roster is a fairly ‘big’ data structure. Rehashing it many times over is not desirable. That’s why we opted to separate them.
If we get rid of the RosterMap and instead store all Rosters in RosterState, it will come with a higher re-hash tax.

I’ve set up a call to discuss these in details tomorrow with @Jasperpotts.

A Roster is considered valid if it satisfies the following conditions:

1. The roster must have at least one RosterEntry.
2. At least one RosterEntry/ies must have a non-zero weight.

> 3. All RosterEntry/ies must have a valid X509Certificate.

Reviewer 8: Maybe we should update the comments on the protobuf to say it is a DER encoded X509Certificate.

Author: I think it already does on lines 148 and 157. I've just made it explicit here too.

4. All RosterEntry/ies must have a valid PairingPublicKey.
5. All RosterEntry/ies must have at least one gossip Endpoint.

> 6. The RosterEntry/ies must be specified in order of ascending node id.

Reviewer 8: Why is this a requirement?

Author: To keep the invariant that exist today in the `AddressBook`. The network topology uses the node ID order to determine what node to/not to connect to. I *think* this invariant is also expected in Reconnect.

Reviewer 8: The code can always sort by ID if it is essential that it does so.

7. All ServiceEndpoint/s must have a valid IP address or domain name (mutually exclusive), and port.
8. The roster must have a unique NodeId for each RosterEntry, and the NodeId key must match the NodeId value in the
   corresponding RosterEntry.

On the submission of a new Candidate Roster, the platform will validate the roster against these conditions.
Note that a constructed `Roster` can be valid, but not accepted by the platform.
For example, if a new candidate roster is set via the API, but its hash evaluates to the hash of the existing

> candidate roster, the new roster will be discarded. That is, the operation has no effect.

Reviewer 8: Practically speaking, this just means the app submitted the same candidate roster twice?

Author: yes. We're only making it explicit that candidate roster submission will be an idempotent operation.

- The `config.txt` file will no longer be used by the platform for storing the address book in existing networks. It
  will no longer be used by the platform code
- Designating `config.txt` introduces inversion of control. It will be at the exclusive prerogative of Services (the
  App) going forward. Platform will no longer be responsible for it, and all the platform code that builds an
  AddressBook from `config.txt` will be removed or refactored and moved into the Services codebase.

The current network Transplant procedure is manual.
DevOps get given a State and `config.txt` file on disk. This is then followed by a software-only upgrade to adopt
the `config.txt` file using that state. This will change to the use of a Genesis Roster.

#### Genesis Roster

![](Proposed%20Network%20Transplant%20flow.drawio.svg)
A `Genesis Roster` is an optional Roster that DevOps may provide to either start a Genesis Network process or Transplant
an existing network. It is essential to distinguish the Genesis Roster from the Candidate or Active rosters.
The Genesis Roster is a special roster used for the sole-purpose of bootstrapping a network. However, its structure is

> exactly the same as the Active or Candidate Rosters.

Reviewer 8: There may be a difference between a "Genesis Roster" (which should only be used at Genesis) and an "Override Roster" (which should override whatever is already in state and override the Genesis Roster). We will need the same things for configs -- genesis, override, and what is in state.

Reviewer 4: It makes sense to name them this way. I have gotten confused with the term "genesis" referring to something that is not related to the genesis of a network.

Reviewer 2: @rbair23, In your mental model of `override roster` is the network starting over from round 1 or trying to continue from the current round?    Does the hashgraph continue seemlessly?   Or is this a disjoint hashgraph?    When the ledger id enters the picture, these overrides will force a reset of the ledger id.

Author: @edward-swirldslabs I think the value is clarity. I agree we should scope the "Genesis Roster" to the case where the network starts at Genesis (round 1 e.g. Genesis Network Process) and an "Override Roster" to the case of a network transplant (round number continues from the last in the state). I'll clarify this in the doc (straightforward diagram change).

Reviewer 8: @edward-swirldslabs I haven't dug into what actions are taken for override. However, we clearly have two different flows -- one where we are at genesis (there is no pre-existing state) and one where we are not at genesis. If we have one override file that behaves differently based on context (i.e. is there a saved state or not), then it becomes exceptionally tricky to work with. The node operator *must* deploy the roster on genesis, and then *must* remove it between the time the network starts and saves a state and the next node restart.

If instead you have two files -- one for genesis, and one for override purposes, then under normal circumstances the node operator either: (a) deploys one for genesis because their node is a genesis node, and then leave that file there for as long as they like (forever even!), or (b) never deploys a file because they're joining an existing network.

But they never have to worry about removing a file between startup and restart. And the only time to use an override file is if there is a **major** problem, or you are moving state to a new network (from mainnet to testnet or something).

Author: I've modified the text and diagrams to detail the need and behaviour of the override roster.

  App) going forward. Platform will no longer be responsible for it, and all the platform code that builds an
  AddressBook from `config.txt` will be removed or refactored and moved into the Services codebase.

### DevOps Workflow changes

The current network Transplant procedure is manual.
DevOps get given a State and `config.txt` file on disk. This is then followed by a software-only upgrade to adopt
the `config.txt` file using that state. This will change to the use of a Genesis Roster.

#### Genesis Roster

A `Genesis Roster` is an optional Roster that DevOps may provide to either start a Genesis Network process or Transplant
an existing network. It is essential to distinguish the Genesis Roster from the Candidate or Active rosters.
The Genesis Roster is a special roster used for the sole-purpose of bootstrapping a network. However, its structure is
exactly the same as the Active or Candidate Rosters.
The equivalent of the Genesis Roster in the current network is the `config.txt` file.
DevOps and Services may choose to create the Genesis Roster from the `config.txt` file or some other mechanism.

> A new method will be added to the `PlatformBuilder` that will be used by the App to set the Genesis Roster.

Reviewer 2: The discussion of the `PlatformBuilder` is impacting of Services not DevOps.   I wonder how much of the description here belongs in the services section.  DevOps is just going to care about artifacts on disk.

Author: I'v added a "What services need to do" summary that helps put the services changes in one place.

In most cases, this Genesis Roster will be null.

```java
//in PlatformBuilder

/*
  Set the Genesis Roster for the network.
  <p>
  This method is used to set the optional Genesis Roster for the network.
  The Genesis Roster is a special roster used for the sole-purpose of bootstrapping a network.
  The Genesis Roster, if set, is immutable and will be used to start a Genesis network Process or Network Transplant process.
  <p>
  @param genesisRoster the Genesis Roster for the network.
 */

> void withGenesisRoster(@NonNull final Roster genesisRoster);

Reviewer 2: Also part of inversion of control is setting the cryptography in the platform builder and the starting state.

Author: That bit is already done - ServicesMain already calls the builder method to set Cryptography (I remember writing this)

```

> The logic to determine whether to invoke this method could be determined by a config flag,

Reviewer 2: I don't think we need to mention the use of a config flag here.   That would be between Services and DevOps.

Author: I want to mention this will require some new logic without the details. I've rephrased it now.

* Network Transplant process == Genesis Roster AND State provided
  When both a Genesis Roster and State are provided, it signifies a network transplant.
  The node will discard any existing candidate roster, rotate the active roster to the previous roster,
  and adopt the provided Genesis Roster as the new Active Roster. This way, the provided network state and Genesis
  Roster are adopted.
  There will be new code required to create the Genesis Roster and pass it to the platform via the `PlatformBuilder` if
  the network is intended to be in a Transplant mode.
  The Services team will be responsible for implementing this, althouth the details are yet to be defined.

* Genesis Network process == Genesis Roster, No State provided
  When a Genesis Roster is provided, but no pre-existing state exists, it will indicate the creation of a new network.
  The node will initiate the TSS key generation process (out of scope for this proposal),
  create a genesis state with the provided Genesis Roster, and start participating in consensus from round 0.

* Keep Network Settings (Normal restart) == No Genesis Roster AND State provided

>   If only a state is provided without a roster, the node will retain its existing network settings, including the active

Reviewer 2: I would indicate in text that if there is a software upgrade, the candidate roster is adopted.

Author: that's been said twice though - but I've added it here again for clarity.

  roster.
  This will be the typical behavior for a node restarting within an established network.

>   ![](Proposed%20Network%20Transplant%20flow.drawio.svg)

Reviewer 4: I like this diagram, but a few things need to be fixed:
1. Add a start and end circle so that all flows end at the end circle.
2. There are three "yes" outputs from the "Genesis Roster Exist?" diamond. There must only be 1

Author: fixed

- Event Validation. The event validation logic uses the address book to determine whether a given event has a valid
  signature. This will be updated to use rosters instead.
- Miscellaneous components. Other components that reference things stored in the address book (like consensus needing
  the node weight or the crypto module verifying keys) will need to be updated to use things stored in rosters instead.
  Some of these components include SelfEventSigner, EventHasher, HealthMonitor, SignedStateSentine, PcesSequencer,
  StateHasher, HashLogger etc.

## Test Plan

Some of the obvious test cases to be covered in the plan include validating one or more of the following scenarios:

1. New valid Candidate Roster created with no subsequent one sent by App. Verify accept.
2. New valid Candidate Roster created with the subsequent one sent by App. Verify accept.
3. Invalid roster(s) sent by the App. Verify reject.
4. Node Failures During Roster Change: What happens if nodes fail or disconnect during a roster change? Verify valid
   node successfully reconnects.
5. Concurrent Roster Updates: What if we make multiple roster updates concurrently? Verify no effect on adoption.

> 6. Roster recovery? The Node receives a candidate roster, crashes. Wake up, reconnect. Verify recovery.

Reviewer 4: This is the same as #4 so it can be removed

Author: they are slightly different test cases. 

#4, the candidate roster didn't get set in the state. But when the node comes back up, if it's in the candidate roster the rest of the network has (now) adopted, it will connect and carry on just fine. 

#6 the candidate roster got set in the state for that node, and does the same test as #4. 

Thinking about it more now, In both, two use cases can be tested; one will be that the nodeId is in the roster, the other will be that it's not. But i think there will probably be more test cases that will be added during implementation.

  <p>
  This method is used to set the optional Genesis Roster for the network.
  The Genesis Roster is a special roster used for the sole-purpose of bootstrapping a network.
  The Genesis Roster, if set, is immutable and will be used to start a Genesis network Process or Network Transplant process.
  <p>
  @param genesisRoster the Genesis Roster for the network.
 */
PlatformBuilder withGenesisRoster(@NonNull final Roster genesisRoster);
```

The logic to determine whether to invoke this method could be determined by some mechanism Services deems appropriate.

#### Override Roster

An `Override Roster` is an optional Roster that DevOps may provide for the explicit purpose of starting a network using
a specific state, with a specific set of nodes, such as during a Network Transplant Process. A common example of
this is Mainnet State being transplanted into a testing network. Provided an existing State and an `Override Roster`,
the platform will adopt that roster.

The next round number for this network will continue from the last round number in the provided state.
The trigger for adopting the `Override Roster` will be the presence of an Override Roster at start-up with a State

> existing on disk. Bear in mind the Network Transplant process is designed primarily with test networks in mind.

Reviewer 4: An upgrade is still required to adopt the override roster because, until we have birth rounds enabled, the software version is the only mechanism by which we determine which roster to use to verify the signatures of events in the PCES written by nodes in the roster being replaced. The consensus node must be able to determine which roster (the override roster or the one originally in the state on disk) to use to verify the signatures of these events. Without a birth round or a software change to make that decision, the verification of PCES will fail.

I do not want to hold this proposal up any further, so please create a ticket to update this proposal after it has been merged and I will resolve this comment and approve.

Author: Given an existing network A, and a test network B that must be transplanted, with A and B running the same software version. Seems like a valid use-case, especially in testing. Will this imply we don't support transplanting A into B _without_ a software version bump, at least until Birth Rounds or DAB (whichever comes first?)?

Reviewer 4: Correct, today the process requires a config-only version upgrade when transplanting into a test network, specifically because of the problem I described.

Author: updated the diagram and text to reflect this.

which there is no existing State, and the round number begins at 1.
However, its structure is exactly the same as the Active or Candidate Rosters.
The equivalent of the Genesis Roster in the current DevOps flow is the `config.txt` file.
DevOps and Services may choose to create the Genesis Roster from the `config.txt` file or some other mechanism.
The trigger for this workflow will be the presence of a Genesis Roster at start-up with no State on disk.

A new method will be added to the `PlatformBuilder` that will be used by the App to set the Genesis Roster.
In most cases, there will be no need to call the method on `PlatformBuilder` to set the Genesis Roster except in the
events we need to bootstrap a Genesis Network.

```java
//in PlatformBuilder

/*
  Set the Genesis Roster for the network.
  <p>
  This method is used to set the optional Genesis Roster for the network.
  The Genesis Roster is a special roster used for the sole-purpose of bootstrapping a network.
  The Genesis Roster, if set, is immutable and will be used to start a Genesis network Process or Network Transplant process.
  <p>
  @param genesisRoster the Genesis Roster for the network.
 */

> void withGenesisRoster(@NonNull final Roster genesisRoster);

Reviewer 7: These builder methods should return `PlatformBuilder`.

DevOps and Services may choose to create the Genesis Roster from the `config.txt` file or some other mechanism.
The trigger for this workflow will be the presence of a Genesis Roster at start-up with no State on disk.

A new method will be added to the `PlatformBuilder` that will be used by the App to set the Genesis Roster.
In most cases, there will be no need to call the method on `PlatformBuilder` to set the Genesis Roster except in the
events we need to bootstrap a Genesis Network.

```java
//in PlatformBuilder

/*
  Set the Genesis Roster for the network.
  <p>
  This method is used to set the optional Genesis Roster for the network.
  The Genesis Roster is a special roster used for the sole-purpose of bootstrapping a network.
  The Genesis Roster, if set, is immutable and will be used to start a Genesis network Process or Network Transplant process.
  <p>
  @param genesisRoster the Genesis Roster for the network.
 */
void withGenesisRoster(@NonNull final Roster genesisRoster);
```

The logic to determine whether to invoke this method could be determined by some mechanism Services deems appropriate.

#### Override Roster

> An `Override Roster` is an optional Roster that DevOps may provide for the explicit purpose of starting a network in a

Reviewer 4: ```suggestion
An `Override Roster` is an optional Roster that DevOps may provide for the explicit purpose of starting a network using a
```

specific state,
with a specific set of nodes, such as during a Network Transplant Process. A common example of this is Mainnet State
being transplanted into Testnet.
Provided an existing State and an `Override Roster`, the platform will adopt that roster.
The next round number for this network will continue from the last round number in the provided state.
The trigger for adopting the `Override Roster` will be the presence of an Override Roster at start-up with a State
existing on disk.
Bear in mind the Network Transplant process is designed primarily with test networks in mind.

> Upon adoption of the Override Roster, the Override Roster will be deleted from disk.

Reviewer 4: ```suggestion
Upon adoption of the Override Roster, the Override Roster should be deleted from disk.
```

In test networks where this process is designed to be used, the Override Roster can be easily accessed and deleted by
DevOps.

A new method will be added to the `PlatformBuilder` that will be used by the App to set the Override Roster.
In most cases, as it's the case for a Genesis Roster, there will be no need to call the method on `PlatformBuilder`
to set the Override Roster except in the events we need to transplant a network.

```java
//in PlatformBuilder
/*
  Set the Override Roster for the network.
  <p>
  This method is used to set the optional Override Roster for the network.
  The Override Roster is a special roster used for the sole purpose of Transplanting an existing network.
  The Override Roster, if set, is immutable and will be used to trigger a Network Transplant process.
  <p>
  @param overrideRoster the Override Roster for the network.
 */
void withOverrideRoster(@NonNull final Roster overrideRoster);
```

#### New Transplant Procedure

The App will decide a network transplant sequence based on the following heuristics:

> Network Transplant process == Override Roster AND State provided

Reviewer 4: I think this process will require a software upgrade as well. There will be events on disk in the PCES that are signed with the previous roster which need to be read into memory and have their signatures validated using the old roster. They way we identify the roster to use to verify PCES signatures is the event version. If the PCES event version is less than the current software version, the old roster is used. If the PCES event version is the same as the current software version, the active roster is used.

Author: updated.

An exception will be thrown if the Roster is invalid.
The durability of the submitted Roster is ensured via storing it in the state.

The existing `PlatformState` class is in the middle of a refactor, and in the long term, may cease to exist.
However, its replacement will continue to have this new API to set a candidate roster.

## Data Structure

The App is already responsible for managing the address book.
We propose that it continues to do so, and when it receives a HAPI transaction, creates a candidate Roster object
and passes it to the platform via the new API. The platform will then validate the roster and store it in the state.

### Protobuf

#### Roster

The protobuf definition for the Roster will look as follows:

```proto
/**
 * A single roster in the network state.
 * <p>
 * The roster SHALL be a PBJ map of key `NodeId`, value `RosterEntry`.
 */
message Roster {

    /**
     * Map of roster entries, one per consensus node.
     * <p>
     * This map SHALL contain roster entries in natural order of ascending node ids.
     * The node Id key SHALL match the node Id in the value.
     * This map SHALL NOT be empty.<br/>
     */

>     map<uint64, RosterEntry> mapInt32ToString = 1;

Reviewer: Minor: PBJ is just an implementation of the protobuf support. In theory, this protobuf definition can be used with other implementations, for example, with Google protoc. I'd nuke the word "PBJ" altogether here. But if you insist on having a qualifier, I'd suggest to replace it with "protobuf" instead.

Also, please change the name of the field to, say, `rosterEntries` or something alike.

Reviewer: Thanks for fixing this. Looks great!

    bytes gossip_ca_certificate = 3;

    /**
     * An ALT_BN128 elliptic curve public encryption key.<br/>
     * <p>
     * The elliptic curve type may change in the future. For example <br/>

>      * If the Ethereum ecosystem creates precompiles for BLS12_381,

Reviewer: Minor nit: `If` capitalization at line 146 is wrong. Also, a comma seems to be missing before it.

     * then endpoints in this list MAY supply either IP address or FQDN, but
     * SHALL NOT supply both values for the same endpoint.<br/>
     * This list SHALL NOT be empty.<br/>
     */
    repeated proto.ServiceEndpoint gossip_endpoint = 5;
}
```

#### Roster Map

> A States API map `Roster Map` (separate from the PBJ Map modeled above) will be introduced. It will contain keys

Reviewer: Minor: there's no longer any PBJ maps whatsoever, so you could nuke the remark.

Immediately following the adoption of a candidate roster, this list will have two elements — the previous active roster
and the current active roster (which was the candidate roster). The choice of a list is suitable here as it is much
cheaper to iterate over 2-3 elements comparing their `long` round numbers than it is to do a hash lookup. When we
reach full Dynamic Address Book, this design choice may need to be revisited.

If a `candidateRosterHash` hash entry already exist (hash collisions) in the map of Rosters, it will be discarded.
That is, setting a candidate roster is an idempotent operation. One benefit of this indirection (using a Map
instead of a Singleton) is that it avoids moving data around in the merkle state. Another benefit is that adoption
trigger becomes straightforward (App sets the roster) with delineated responsibilities between the App and the platform.

### Roster Validity

A Roster is considered valid if it satisfies the following conditions:

1. The roster must have at least one RosterEntry.
2. At least one RosterEntry/ies must have a non-zero weight.
3. All RosterEntry/ies must have a valid gossip_ca_certificate which is a DER encoded X509Certificate.

> 4. All RosterEntry/ies must have a valid tss_encryption_key which is a DER encoded X509Certificate.

Reviewer: The tss_encryption_key isn't a DER encoded X509Certificate. To avoid duplicating the format descriptions, I suggest to rephrase the validation checks 3 and 4 to simply refer to the protobuf specifications where the format is actually specified.

5. All RosterEntry/ies must have at least one gossip Endpoint.
6. All ServiceEndpoint/s must have a valid IP address or domain name (mutually exclusive), and port.
7. The roster must have a unique NodeId for each RosterEntry, and the NodeId key must match the NodeId value in the

>    corresponding RosterEntry.

Reviewer: Since it's no longer a map, the check 7 needs to be rephrased. It now needs to validate that the RosterEntry list doesn't contain entries with duplicate NodeIds.

### On Disk

There will continue to be components used in creating the Roster - such as the private EC (elliptic curve) key, the RSA
private key and the signing X509Certificate certificate - that will be stored on disk. However, it is at the
discretion of the Services team to manage the lifecycle of these files, and not the platform. There will not be any
other separate artifacts from this proposal stored directly on disk.

## Services and DevOps changes (Inversion of control)

The current startup procedure will be altered as follows:

- The `config.txt` file will no longer be used by the platform for storing the address book in existing networks. It
  will no longer be used by the platform code
- Designating `config.txt` introduces inversion of control. It will be at the exclusive prerogative of Services (the
  App) going forward. Platform will no longer be responsible for it, and all the platform code that builds an
  AddressBook from `config.txt` will be removed or refactored and moved into the Services codebase.

> The current network Transplant procedure is manual. DevOps get given a State and `config.txt` file on disk. This is

Reviewer 4: ```suggestion
The current network Transplant procedure is manual. DevOps is given a State and `config.txt` file on disk. This is
```

> then followed by a software-only upgrade to adopt the `config.txt` file using that state. This will change to the

Reviewer 4: ```suggestion
then followed by a config-only upgrade to adopt the `config.txt` file using that state. This will change to the
```

#### Override Roster

An `Override Roster` is an optional Roster that DevOps may provide for the explicit purpose of starting a network using

> a

Reviewer 2: spacing, this line has an `a` by itself.
